{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setting up"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "bat"
    }
   },
   "source": [
    "Before starting, we need to download the model weights for TamGen. This might take a while... Just run the cells below to get started\n",
    "\n",
    "Instructions for manual install:\n",
    "\n",
    "Download link:\n",
    "https://zenodo.org/records/13751391\n",
    "\n",
    "- Download the checkpoints and extract them in the root folder\n",
    "- Copy 'checkpoints/crossdocked_model/checkpoint_best.pt' and place it in the 'gpt_model' model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!wget -nc https://zenodo.org/records/13751391/files/checkpoints.zip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!unzip checkpoints.zip -d ./"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!cp ./checkpoints/crossdocked_model/checkpoint_best.pt ./gpt_model/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pre-Processing\n",
    "\n",
    "We create a KG in NetworkX-arangoDB\n",
    "\n",
    "This requires downloading multi-modality datasets from [Bio Snap](https://snap.stanford.edu/biodata/) and other data source. This notebook does simple data engineering before pushing into ArangoDB to create a fully connected network.\n",
    "\n",
    "##### Modalities:\n",
    "- Drug-Drug interaction with side-effects\n",
    "- Drug-Gene interaction\n",
    "- Disease-Drug interaction\n",
    "- Disease-Disease interaction\n",
    "- Disease-Function interaction\n",
    "- Function-Function interaction\n",
    "- Gene-Function interaction\n",
    "- Gene-Gene interaction\n",
    "- Gene-Protien interaction\n",
    "- Genomic Region-Genomic Region interaction\n",
    "- Protien-Protien interaction\n",
    "- Protien-Protien-Tissue interaction\n",
    "- Tissue-Function-Gene interaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File ‘./data/ChCh-Miner_durgbank-chem-chem.tsv.gz’ already there; not retrieving.\n",
      "\n",
      "File ‘./data/ChG-Miner_miner-chem-gene.tsv.gz’ already there; not retrieving.\n",
      "\n",
      "File ‘./data/DCh-Miner_miner-disease-chemical.tsv.gz’ already there; not retrieving.\n",
      "\n",
      "File ‘./data/DD-Miner_miner-disease-disease.tsv.gz’ already there; not retrieving.\n",
      "\n",
      "File ‘./data/DF-Miner_miner-disease-function.tsv.gz’ already there; not retrieving.\n",
      "\n",
      "File ‘./data/DG-Miner_miner-disease-gene.tsv.gz’ already there; not retrieving.\n",
      "\n",
      "File ‘./data/FF-Miner_miner-func-func.tsv.gz’ already there; not retrieving.\n",
      "\n",
      "File ‘./data/GF-Miner_miner-gene-function.tsv.gz’ already there; not retrieving.\n",
      "\n",
      "File ‘./data/GP-Miner_miner-gene-protein.tsv.gz’ already there; not retrieving.\n",
      "\n",
      "File ‘./data/chembl_35_chemreps.txt.gz’ already there; not retrieving.\n",
      "\n",
      "File ‘./data/D-DoMiner_miner-diseaseDOID.tsv.gz’ already there; not retrieving.\n",
      "\n",
      "File ‘./data/D-MeshMiner_miner-disease.tsv.gz’ already there; not retrieving.\n",
      "\n",
      "File ‘./data/D-OmimMiner_miner-diseaseOMIM.tsv.gz’ already there; not retrieving.\n",
      "\n",
      "File ‘./data/G-SynMiner_miner-geneHUGO.tsv.gz’ already there; not retrieving.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Interactions\n",
    "!wget -nc -P ./data/ https://snap.stanford.edu/biodata/datasets/10001/files/ChCh-Miner_durgbank-chem-chem.tsv.gz\n",
    "!wget -nc -P ./data/ https://snap.stanford.edu/biodata/datasets/10002/files/ChG-Miner_miner-chem-gene.tsv.gz\n",
    "!wget -nc -P ./data/ https://snap.stanford.edu/biodata/datasets/10004/files/DCh-Miner_miner-disease-chemical.tsv.gz\n",
    "!wget -nc -P ./data/ https://snap.stanford.edu/biodata/datasets/10006/files/DD-Miner_miner-disease-disease.tsv.gz\n",
    "!wget -nc -P ./data/ https://snap.stanford.edu/biodata/datasets/10019/files/DF-Miner_miner-disease-function.tsv.gz\n",
    "!wget -nc -P ./data/ https://snap.stanford.edu/biodata/datasets/10020/files/DG-Miner_miner-disease-gene.tsv.gz\n",
    "!wget -nc -P ./data/ https://snap.stanford.edu/biodata/datasets/10026/files/FF-Miner_miner-func-func.tsv.gz\n",
    "!wget -nc -P ./data/ https://snap.stanford.edu/biodata/datasets/10024/files/GF-Miner_miner-gene-function.tsv.gz\n",
    "!wget -nc -P ./data/ https://snap.stanford.edu/biodata/datasets/10027/files/GP-Miner_miner-gene-protein.tsv.gz\n",
    "\n",
    "# Mappings\n",
    "!wget -nc -P ./data/ https://ftp.ebi.ac.uk/pub/databases/chembl/ChEMBLdb/latest/chembl_35_chemreps.txt.gz\n",
    "\n",
    "# Entities and Feature tables\n",
    "!wget -nc -P ./data/ https://snap.stanford.edu/biodata/datasets/10021/files/D-DoMiner_miner-diseaseDOID.tsv.gz\n",
    "!wget -nc -P ./data/ https://snap.stanford.edu/biodata/datasets/10003/files/D-MeshMiner_miner-disease.tsv.gz\n",
    "!wget -nc -P ./data/ https://snap.stanford.edu/biodata/datasets/10025/files/D-OmimMiner_miner-diseaseOMIM.tsv.gz\n",
    "!wget -nc -P ./data/ https://snap.stanford.edu/biodata/datasets/10022/files/G-SynMiner_miner-geneHUGO.tsv.gz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import requests\n",
    "import ast\n",
    "import json\n",
    "import hashlib\n",
    "from datetime import datetime\n",
    "from glob import glob\n",
    "from io import StringIO\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from dotenv import load_dotenv\n",
    "from arango import ArangoClient\n",
    "from biomart import BiomartServer\n",
    "\n",
    "from transformers import AutoTokenizer, AutoModel\n",
    "import torch\n",
    "\n",
    "from langgraph.prebuilt import create_react_agent\n",
    "from langgraph.checkpoint.memory import MemorySaver\n",
    "from langchain_openai import ChatOpenAI\n",
    "from langchain_community.graphs import ArangoGraph\n",
    "from langchain_community.chains.graph_qa.arangodb import ArangoGraphQAChain\n",
    "from langchain_core.tools import tool\n",
    "\n",
    "from DeepPurpose import utils\n",
    "from DeepPurpose import DTI as models\n",
    "\n",
    "from rdkit import Chem, DataStructs\n",
    "from rdkit.Chem import MACCSkeys\n",
    "from rdkit.Chem import Draw, AllChem\n",
    "\n",
    "from Bio.PDB import MMCIFParser\n",
    "\n",
    "import faiss"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading Dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Entities and Feature Tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>doid</th>\n",
       "      <th>name</th>\n",
       "      <th>definition</th>\n",
       "      <th>synonym</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DOID:0001816</td>\n",
       "      <td>angiosarcoma</td>\n",
       "      <td>A malignant vascular tumor that results_in rap...</td>\n",
       "      <td>hemangiosarcoma EXACT []</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DOID:0002116</td>\n",
       "      <td>pterygium</td>\n",
       "      <td>NaN</td>\n",
       "      <td>surfer's eye EXACT []</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DOID:0014667</td>\n",
       "      <td>disease of metabolism</td>\n",
       "      <td>A disease that involving errors in metabolic p...</td>\n",
       "      <td>metabolic disease  EXACT [SNOMEDCT_2005_07_31:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DOID:0050001</td>\n",
       "      <td>Actinomadura madurae infectious disease</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DOID:0050002</td>\n",
       "      <td>Actinomadura pelletieri infectious disease</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9242</th>\n",
       "      <td>DOID:9989</td>\n",
       "      <td>metastasis to the orbit</td>\n",
       "      <td>NaN</td>\n",
       "      <td>secondary malignant neoplasm of orbit (disorde...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9243</th>\n",
       "      <td>DOID:999</td>\n",
       "      <td>eosinophilia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eosinophilic leukocytosis EXACT [MTHICD9_2006:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9244</th>\n",
       "      <td>DOID:9993</td>\n",
       "      <td>hypoglycemia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hypoglycaemia EXACT [SNOMEDCT_2005_07_31:15469...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9245</th>\n",
       "      <td>DOID:9995</td>\n",
       "      <td>endocrine and metabolic disturbances specific ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9246</th>\n",
       "      <td>DOID:9997</td>\n",
       "      <td>peripartum cardiomyopathy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>postpartum peripartum cardiomyopathy EXACT []</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9247 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              doid                                               name  \\\n",
       "0     DOID:0001816                                       angiosarcoma   \n",
       "1     DOID:0002116                                          pterygium   \n",
       "2     DOID:0014667                              disease of metabolism   \n",
       "3     DOID:0050001            Actinomadura madurae infectious disease   \n",
       "4     DOID:0050002         Actinomadura pelletieri infectious disease   \n",
       "...            ...                                                ...   \n",
       "9242     DOID:9989                            metastasis to the orbit   \n",
       "9243      DOID:999                                       eosinophilia   \n",
       "9244     DOID:9993                                       hypoglycemia   \n",
       "9245     DOID:9995  endocrine and metabolic disturbances specific ...   \n",
       "9246     DOID:9997                          peripartum cardiomyopathy   \n",
       "\n",
       "                                             definition  \\\n",
       "0     A malignant vascular tumor that results_in rap...   \n",
       "1                                                   NaN   \n",
       "2     A disease that involving errors in metabolic p...   \n",
       "3                                                   NaN   \n",
       "4                                                   NaN   \n",
       "...                                                 ...   \n",
       "9242                                                NaN   \n",
       "9243                                                NaN   \n",
       "9244                                                NaN   \n",
       "9245                                                NaN   \n",
       "9246                                                NaN   \n",
       "\n",
       "                                                synonym  \n",
       "0                              hemangiosarcoma EXACT []  \n",
       "1                                 surfer's eye EXACT []  \n",
       "2     metabolic disease  EXACT [SNOMEDCT_2005_07_31:...  \n",
       "3                                                   NaN  \n",
       "4                                                   NaN  \n",
       "...                                                 ...  \n",
       "9242  secondary malignant neoplasm of orbit (disorde...  \n",
       "9243  Eosinophilic leukocytosis EXACT [MTHICD9_2006:...  \n",
       "9244  Hypoglycaemia EXACT [SNOMEDCT_2005_07_31:15469...  \n",
       "9245                                                NaN  \n",
       "9246      postpartum peripartum cardiomyopathy EXACT []  \n",
       "\n",
       "[9247 rows x 4 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Disease DOID Synopses\n",
    "\n",
    "Doid = pd.read_csv(\n",
    "    \"./data/D-DoMiner_miner-diseaseDOID.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    header=0,\n",
    "    names=[\"doid\", \"name\", \"definition\", \"synonym\"],\n",
    ")\n",
    "\n",
    "Doid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mesh</th>\n",
       "      <th>name</th>\n",
       "      <th>definition</th>\n",
       "      <th>synonym</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MESH:C538288</td>\n",
       "      <td>10p Deletion Syndrome (Partial)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Chromosome 10, 10p- Partial|Chromosome 10, mon...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MESH:C535484</td>\n",
       "      <td>13q deletion syndrome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Chromosome 13q deletion|Chromosome 13q deletio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MESH:C579849</td>\n",
       "      <td>15q24 Microdeletion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15q24 Deletion|15q24 Microdeletion Syndrome|In...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MESH:C579850</td>\n",
       "      <td>16p11.2 Deletion Syndrome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MESH:C567076</td>\n",
       "      <td>17,20-Lyase Deficiency, Isolated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17-Alpha-Hydroxylase-17,20-Lyase Deficiency, C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11327</th>\n",
       "      <td>MESH:C536729</td>\n",
       "      <td>Zunich neuroectodermal syndrome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHIME syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11328</th>\n",
       "      <td>MESH:C536730</td>\n",
       "      <td>Zuska's Disease</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lactation and squamous metaplasia of lactifero...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11329</th>\n",
       "      <td>MESH:C565223</td>\n",
       "      <td>Zygodactyly 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11330</th>\n",
       "      <td>MESH:D015051</td>\n",
       "      <td>Zygomatic Fractures</td>\n",
       "      <td>Fractures of the zygoma.</td>\n",
       "      <td>Fractures, Zygomatic|Fracture, Zygomatic|Zygom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11331</th>\n",
       "      <td>MESH:D020096</td>\n",
       "      <td>Zygomycosis</td>\n",
       "      <td>Infection in humans and animals caused by fung...</td>\n",
       "      <td>Entomophthoramycoses|Entomophthoramycosis|Phyc...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11332 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               mesh                              name  \\\n",
       "0      MESH:C538288   10p Deletion Syndrome (Partial)   \n",
       "1      MESH:C535484             13q deletion syndrome   \n",
       "2      MESH:C579849               15q24 Microdeletion   \n",
       "3      MESH:C579850         16p11.2 Deletion Syndrome   \n",
       "4      MESH:C567076  17,20-Lyase Deficiency, Isolated   \n",
       "...             ...                               ...   \n",
       "11327  MESH:C536729   Zunich neuroectodermal syndrome   \n",
       "11328  MESH:C536730                   Zuska's Disease   \n",
       "11329  MESH:C565223                     Zygodactyly 1   \n",
       "11330  MESH:D015051               Zygomatic Fractures   \n",
       "11331  MESH:D020096                       Zygomycosis   \n",
       "\n",
       "                                              definition  \\\n",
       "0                                                    NaN   \n",
       "1                                                    NaN   \n",
       "2                                                    NaN   \n",
       "3                                                    NaN   \n",
       "4                                                    NaN   \n",
       "...                                                  ...   \n",
       "11327                                                NaN   \n",
       "11328                                                NaN   \n",
       "11329                                                NaN   \n",
       "11330                           Fractures of the zygoma.   \n",
       "11331  Infection in humans and animals caused by fung...   \n",
       "\n",
       "                                                 synonym  \n",
       "0      Chromosome 10, 10p- Partial|Chromosome 10, mon...  \n",
       "1      Chromosome 13q deletion|Chromosome 13q deletio...  \n",
       "2      15q24 Deletion|15q24 Microdeletion Syndrome|In...  \n",
       "3                                                    NaN  \n",
       "4      17-Alpha-Hydroxylase-17,20-Lyase Deficiency, C...  \n",
       "...                                                  ...  \n",
       "11327                                     CHIME syndrome  \n",
       "11328  Lactation and squamous metaplasia of lactifero...  \n",
       "11329                                                ZD1  \n",
       "11330  Fractures, Zygomatic|Fracture, Zygomatic|Zygom...  \n",
       "11331  Entomophthoramycoses|Entomophthoramycosis|Phyc...  \n",
       "\n",
       "[11332 rows x 4 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Disease MESH Synopses\n",
    "\n",
    "Mesh = pd.read_csv(\n",
    "    \"./data/D-MeshMiner_miner-disease.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    header=0,\n",
    "    names=[\"mesh\", \"name\", \"definition\", \"synonym\"],\n",
    ")\n",
    "\n",
    "Mesh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>omim</th>\n",
       "      <th>phenotypes</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene</th>\n",
       "      <th>location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OMIM:115665</td>\n",
       "      <td>Cataract 8, multiple types (2)</td>\n",
       "      <td>Cataract, congenital, Volkmann type</td>\n",
       "      <td>CTRCT8, CCV</td>\n",
       "      <td>1pter-p36.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OMIM:607671</td>\n",
       "      <td>Dystonia 13, torsion (2)</td>\n",
       "      <td>Dystonia 13, torsion</td>\n",
       "      <td>DYT13</td>\n",
       "      <td>1p36.32-p36.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OMIM:606242</td>\n",
       "      <td>Kondoh syndrome (2)</td>\n",
       "      <td>Kondoh syndrome (mental retardation, microceph...</td>\n",
       "      <td>KONDS</td>\n",
       "      <td>1p36.32-p35.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OMIM:614414</td>\n",
       "      <td>Deafness, autosomal recessive 96 (2)</td>\n",
       "      <td>Deafness, autosomal recessive 96</td>\n",
       "      <td>DFNB96</td>\n",
       "      <td>1p36.31-p36.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OMIM:609918</td>\n",
       "      <td>Gallbladder disease 2 (2)</td>\n",
       "      <td>Gallbladder disease 2</td>\n",
       "      <td>GBD2</td>\n",
       "      <td>1p36.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1186</th>\n",
       "      <td>OMIM:300519</td>\n",
       "      <td>Mental retardation, X-linked, syndromic, Marti...</td>\n",
       "      <td>Mental retardation, X-linked, syndromic, Marti...</td>\n",
       "      <td>MRXSMP</td>\n",
       "      <td>Chr.X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1187</th>\n",
       "      <td>OMIM:400042</td>\n",
       "      <td>Spermatogenic failure, Y-linked, 1 (4)</td>\n",
       "      <td>Chromosome Yq11 interstitial deletion syndrome</td>\n",
       "      <td>DELYq11, CYDELq11, SPGFY1</td>\n",
       "      <td>Yq11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1188</th>\n",
       "      <td>OMIM:475000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Growth control, Y-chromosome influenced</td>\n",
       "      <td>GCY, TSY, STA</td>\n",
       "      <td>Yq12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1189</th>\n",
       "      <td>OMIM:400043</td>\n",
       "      <td>Deafness, Y-linked 1 (1)</td>\n",
       "      <td>Deafness, Y-linked 1</td>\n",
       "      <td>DFNY1</td>\n",
       "      <td>Chr.Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1190</th>\n",
       "      <td>OMIM:400004</td>\n",
       "      <td>Retinitis pigmentosa, Y-linked (2)</td>\n",
       "      <td>Retinitis pigmentosa, Y-linked</td>\n",
       "      <td>RPY</td>\n",
       "      <td>Chr.Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1191 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             omim                                         phenotypes  \\\n",
       "0     OMIM:115665                     Cataract 8, multiple types (2)   \n",
       "1     OMIM:607671                           Dystonia 13, torsion (2)   \n",
       "2     OMIM:606242                                Kondoh syndrome (2)   \n",
       "3     OMIM:614414               Deafness, autosomal recessive 96 (2)   \n",
       "4     OMIM:609918                          Gallbladder disease 2 (2)   \n",
       "...           ...                                                ...   \n",
       "1186  OMIM:300519  Mental retardation, X-linked, syndromic, Marti...   \n",
       "1187  OMIM:400042             Spermatogenic failure, Y-linked, 1 (4)   \n",
       "1188  OMIM:475000                                                NaN   \n",
       "1189  OMIM:400043                           Deafness, Y-linked 1 (1)   \n",
       "1190  OMIM:400004                 Retinitis pigmentosa, Y-linked (2)   \n",
       "\n",
       "                                              gene_name  \\\n",
       "0                   Cataract, congenital, Volkmann type   \n",
       "1                                  Dystonia 13, torsion   \n",
       "2     Kondoh syndrome (mental retardation, microceph...   \n",
       "3                      Deafness, autosomal recessive 96   \n",
       "4                                 Gallbladder disease 2   \n",
       "...                                                 ...   \n",
       "1186  Mental retardation, X-linked, syndromic, Marti...   \n",
       "1187     Chromosome Yq11 interstitial deletion syndrome   \n",
       "1188            Growth control, Y-chromosome influenced   \n",
       "1189                               Deafness, Y-linked 1   \n",
       "1190                     Retinitis pigmentosa, Y-linked   \n",
       "\n",
       "                           gene        location  \n",
       "0                   CTRCT8, CCV    1pter-p36.13  \n",
       "1                         DYT13  1p36.32-p36.13  \n",
       "2                         KONDS   1p36.32-p35.3  \n",
       "3                        DFNB96  1p36.31-p36.13  \n",
       "4                          GBD2         1p36.21  \n",
       "...                         ...             ...  \n",
       "1186                     MRXSMP           Chr.X  \n",
       "1187  DELYq11, CYDELq11, SPGFY1            Yq11  \n",
       "1188              GCY, TSY, STA            Yq12  \n",
       "1189                      DFNY1           Chr.Y  \n",
       "1190                        RPY           Chr.Y  \n",
       "\n",
       "[1191 rows x 5 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# OMIM Genetic Disorders\n",
    "\n",
    "Omim = pd.read_csv(\n",
    "    \"./data/D-OmimMiner_miner-diseaseOMIM.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    header=0,\n",
    "    names=[\"omim\", \"phenotypes\", \"gene_name\", \"gene\", \"location\", \"_\"],\n",
    ")\n",
    "\n",
    "Omim = Omim.iloc[:, :-1]\n",
    "\n",
    "Omim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/xb/gzq2kh212g9g01hg900w3y1r0000gn/T/ipykernel_55882/2109625620.py:3: DtypeWarning: Columns (32,34,38) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  Gene = pd.read_csv(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensg</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>locus_group</th>\n",
       "      <th>locus_type</th>\n",
       "      <th>status</th>\n",
       "      <th>location</th>\n",
       "      <th>location_sortable</th>\n",
       "      <th>alias_symbol</th>\n",
       "      <th>...</th>\n",
       "      <th>horde_id</th>\n",
       "      <th>merops</th>\n",
       "      <th>imgt</th>\n",
       "      <th>iuphar</th>\n",
       "      <th>kznf_gene_catalog</th>\n",
       "      <th>mamit-trnadb</th>\n",
       "      <th>cd</th>\n",
       "      <th>lncrnadb</th>\n",
       "      <th>enzyme_id</th>\n",
       "      <th>intermediate_filament_db</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>I43.950</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000268895</td>\n",
       "      <td>HGNC:37133</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>A1BG antisense RNA 1</td>\n",
       "      <td>non-coding RNA</td>\n",
       "      <td>RNA, long non-coding</td>\n",
       "      <td>Approved</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>FLJ23569</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>HGNC:24086</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>10q21.1</td>\n",
       "      <td>10q21.1</td>\n",
       "      <td>ACF|ASP|ACF64|ACF65|APOBEC1CF</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>FWP007|S863-7|CPAMD5</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>I39.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000245105</td>\n",
       "      <td>HGNC:27057</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>A2M antisense RNA 1 (head to head)</td>\n",
       "      <td>non-coding RNA</td>\n",
       "      <td>RNA, long non-coding</td>\n",
       "      <td>Approved</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35649</th>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>HGNC:25820</td>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>zyg-11 family member B, cell cycle regulator</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1p32.3</td>\n",
       "      <td>01p32.3</td>\n",
       "      <td>FLJ13456</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35650</th>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>HGNC:13200</td>\n",
       "      <td>ZYX</td>\n",
       "      <td>zyxin</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>7q32</td>\n",
       "      <td>07q32</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35651</th>\n",
       "      <td>ENSG00000274572</td>\n",
       "      <td>HGNC:51695</td>\n",
       "      <td>ZYXP1</td>\n",
       "      <td>zyxin pseudogene 1</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>Approved</td>\n",
       "      <td>8q24.23</td>\n",
       "      <td>08q24.23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35652</th>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>HGNC:29027</td>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>zinc finger ZZ-type and EF-hand domain contain...</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>17p13.3</td>\n",
       "      <td>17p13.3</td>\n",
       "      <td>KIAA0399|ZZZ4|FLJ10821</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35653</th>\n",
       "      <td>ENSG00000036549</td>\n",
       "      <td>HGNC:24523</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>zinc finger ZZ-type containing 3</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1p31.1</td>\n",
       "      <td>01p31.1</td>\n",
       "      <td>DKFZP564I052|ATAC1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35654 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  ensg     hgnc_id      gene  \\\n",
       "0      ENSG00000121410      HGNC:5      A1BG   \n",
       "1      ENSG00000268895  HGNC:37133  A1BG-AS1   \n",
       "2      ENSG00000148584  HGNC:24086      A1CF   \n",
       "3      ENSG00000175899      HGNC:7       A2M   \n",
       "4      ENSG00000245105  HGNC:27057   A2M-AS1   \n",
       "...                ...         ...       ...   \n",
       "35649  ENSG00000162378  HGNC:25820    ZYG11B   \n",
       "35650  ENSG00000159840  HGNC:13200       ZYX   \n",
       "35651  ENSG00000274572  HGNC:51695     ZYXP1   \n",
       "35652  ENSG00000074755  HGNC:29027     ZZEF1   \n",
       "35653  ENSG00000036549  HGNC:24523      ZZZ3   \n",
       "\n",
       "                                               gene_name          locus_group  \\\n",
       "0                                 alpha-1-B glycoprotein  protein-coding gene   \n",
       "1                                   A1BG antisense RNA 1       non-coding RNA   \n",
       "2                         APOBEC1 complementation factor  protein-coding gene   \n",
       "3                                  alpha-2-macroglobulin  protein-coding gene   \n",
       "4                     A2M antisense RNA 1 (head to head)       non-coding RNA   \n",
       "...                                                  ...                  ...   \n",
       "35649       zyg-11 family member B, cell cycle regulator  protein-coding gene   \n",
       "35650                                              zyxin  protein-coding gene   \n",
       "35651                                 zyxin pseudogene 1           pseudogene   \n",
       "35652  zinc finger ZZ-type and EF-hand domain contain...  protein-coding gene   \n",
       "35653                   zinc finger ZZ-type containing 3  protein-coding gene   \n",
       "\n",
       "                      locus_type    status  location location_sortable  \\\n",
       "0      gene with protein product  Approved  19q13.43          19q13.43   \n",
       "1           RNA, long non-coding  Approved  19q13.43          19q13.43   \n",
       "2      gene with protein product  Approved   10q21.1           10q21.1   \n",
       "3      gene with protein product  Approved  12p13.31          12p13.31   \n",
       "4           RNA, long non-coding  Approved  12p13.31          12p13.31   \n",
       "...                          ...       ...       ...               ...   \n",
       "35649  gene with protein product  Approved    1p32.3           01p32.3   \n",
       "35650  gene with protein product  Approved      7q32             07q32   \n",
       "35651                 pseudogene  Approved   8q24.23          08q24.23   \n",
       "35652  gene with protein product  Approved   17p13.3           17p13.3   \n",
       "35653  gene with protein product  Approved    1p31.1           01p31.1   \n",
       "\n",
       "                        alias_symbol  ... horde_id   merops imgt iuphar  \\\n",
       "0                                NaN  ...      NaN  I43.950  NaN    NaN   \n",
       "1                           FLJ23569  ...      NaN      NaN  NaN    NaN   \n",
       "2      ACF|ASP|ACF64|ACF65|APOBEC1CF  ...      NaN      NaN  NaN    NaN   \n",
       "3               FWP007|S863-7|CPAMD5  ...      NaN  I39.001  NaN    NaN   \n",
       "4                                NaN  ...      NaN      NaN  NaN    NaN   \n",
       "...                              ...  ...      ...      ...  ...    ...   \n",
       "35649                       FLJ13456  ...      NaN      NaN  NaN    NaN   \n",
       "35650                            NaN  ...      NaN      NaN  NaN    NaN   \n",
       "35651                            NaN  ...      NaN      NaN  NaN    NaN   \n",
       "35652         KIAA0399|ZZZ4|FLJ10821  ...      NaN      NaN  NaN    NaN   \n",
       "35653             DKFZP564I052|ATAC1  ...      NaN      NaN  NaN    NaN   \n",
       "\n",
       "      kznf_gene_catalog mamit-trnadb   cd lncrnadb enzyme_id  \\\n",
       "0                   NaN          NaN  NaN      NaN       NaN   \n",
       "1                   NaN          NaN  NaN      NaN       NaN   \n",
       "2                   NaN          NaN  NaN      NaN       NaN   \n",
       "3                   NaN          NaN  NaN      NaN       NaN   \n",
       "4                   NaN          NaN  NaN      NaN       NaN   \n",
       "...                 ...          ...  ...      ...       ...   \n",
       "35649               NaN          NaN  NaN      NaN       NaN   \n",
       "35650               NaN          NaN  NaN      NaN       NaN   \n",
       "35651               NaN          NaN  NaN      NaN       NaN   \n",
       "35652               NaN          NaN  NaN      NaN       NaN   \n",
       "35653               NaN          NaN  NaN      NaN       NaN   \n",
       "\n",
       "       intermediate_filament_db  \n",
       "0                           NaN  \n",
       "1                           NaN  \n",
       "2                           NaN  \n",
       "3                           NaN  \n",
       "4                           NaN  \n",
       "...                         ...  \n",
       "35649                       NaN  \n",
       "35650                       NaN  \n",
       "35651                       NaN  \n",
       "35652                       NaN  \n",
       "35653                       NaN  \n",
       "\n",
       "[35654 rows x 48 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Gene\n",
    "\n",
    "Gene = pd.read_csv(\n",
    "    \"./data/G-SynMiner_miner-geneHUGO.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "\n",
    "Gene = Gene.rename(columns={'# ensembl_gene_id': 'ensg', 'symbol': 'gene', 'name': 'gene_name'})\n",
    "\n",
    "Gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "tokenizer = AutoTokenizer.from_pretrained(\"seyonec/ChemBERTa-zinc-base-v1\")\n",
    "model = AutoModel.from_pretrained(\"seyonec/ChemBERTa-zinc-base-v1\")\n",
    "\n",
    "def get_chemberta_embedding(smiles):\n",
    "    \"\"\"Generate a ChemBERTa embedding for a molecule, ensuring input is a string.\"\"\"\n",
    "    if not isinstance(smiles, str) or not smiles.strip():\n",
    "        return None \n",
    "\n",
    "    inputs = tokenizer(smiles, return_tensors=\"pt\", padding=True, truncation=True)\n",
    "    with torch.no_grad():\n",
    "        outputs = model(**inputs)\n",
    "    return outputs.last_hidden_state.mean(dim=1).tolist()[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This script processes DrugBank and ChEMBL chemical data to generate molecular embeddings using a ChemBERTa model. The final dataset contains drugs with their molecular representations and corresponding vector embeddings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug</th>\n",
       "      <th>accession</th>\n",
       "      <th>drug_name</th>\n",
       "      <th>cas</th>\n",
       "      <th>unii</th>\n",
       "      <th>synonym</th>\n",
       "      <th>key</th>\n",
       "      <th>chembl</th>\n",
       "      <th>smiles</th>\n",
       "      <th>inchi</th>\n",
       "      <th>generated</th>\n",
       "      <th>embedding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DB00014</td>\n",
       "      <td>BTD00113 | BIOD00113</td>\n",
       "      <td>Goserelin</td>\n",
       "      <td>65807-02-5</td>\n",
       "      <td>0F65R8P09N</td>\n",
       "      <td>Goserelin | Goserelina</td>\n",
       "      <td>BLCLNMBMMGCOAS-URPVMXJPSA-N</td>\n",
       "      <td>CHEMBL1201247</td>\n",
       "      <td>CC(C)C[C@H](NC(=O)[C@@H](COC(C)(C)C)NC(=O)[C@H...</td>\n",
       "      <td>InChI=1S/C59H84N18O14/c1-31(2)22-40(49(82)68-3...</td>\n",
       "      <td>False</td>\n",
       "      <td>[0.26175203919410706, 0.019564686343073845, -0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DB00027</td>\n",
       "      <td>BTD00036 | BIOD00036</td>\n",
       "      <td>Gramicidin D</td>\n",
       "      <td>1405-97-6</td>\n",
       "      <td>5IE62321P4</td>\n",
       "      <td>Bacillus brevis gramicidin D | Gramicidin | Gr...</td>\n",
       "      <td>NDAYQJDHGXTBJL-MWWSRJDJSA-N</td>\n",
       "      <td>CHEMBL557217</td>\n",
       "      <td>CC(C)C[C@@H](NC(=O)CNC(=O)[C@@H](NC=O)C(C)C)C(...</td>\n",
       "      <td>InChI=1S/C96H135N19O16/c1-50(2)36-71(105-79(11...</td>\n",
       "      <td>False</td>\n",
       "      <td>[0.505556583404541, 0.0775429829955101, -0.058...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DB00035</td>\n",
       "      <td>BTD00112 | BTD00061 | BIOD00112 | BIOD00061</td>\n",
       "      <td>Desmopressin</td>\n",
       "      <td>16679-58-6</td>\n",
       "      <td>ENR1LLB0FP</td>\n",
       "      <td>1-(3-mercaptopropionic acid)-8-D-arginine-vaso...</td>\n",
       "      <td>NFLWUMRGJYTJIN-PNIOQBSNSA-N</td>\n",
       "      <td>CHEMBL1429</td>\n",
       "      <td>N=C(N)NCCC[C@@H](NC(=O)[C@@H]1CCCN1C(=O)[C@@H]...</td>\n",
       "      <td>InChI=1S/C46H64N14O12S2/c47-35(62)15-14-29-40(...</td>\n",
       "      <td>False</td>\n",
       "      <td>[0.1573418825864792, -0.010056326165795326, -0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DB00050</td>\n",
       "      <td>BTD00115 | APRD00686 | BIOD00115</td>\n",
       "      <td>Cetrorelix</td>\n",
       "      <td>120287-85-6</td>\n",
       "      <td>OON1HFZ4BA</td>\n",
       "      <td>Cetrorelix | Cetrorelixum</td>\n",
       "      <td>SBNPWPIBESPSIF-MHWMIDJBSA-N</td>\n",
       "      <td>CHEMBL1200490</td>\n",
       "      <td>CC(=O)N[C@H](Cc1ccc2ccccc2c1)C(=O)N[C@H](Cc1cc...</td>\n",
       "      <td>InChI=1S/C70H92ClN17O14/c1-39(2)31-52(61(94)82...</td>\n",
       "      <td>False</td>\n",
       "      <td>[0.24220487475395203, 0.14190606772899628, -0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DB00080</td>\n",
       "      <td>BTD00111 | BIOD00111</td>\n",
       "      <td>Daptomycin</td>\n",
       "      <td>103060-53-3</td>\n",
       "      <td>NWQ5N31VKK</td>\n",
       "      <td>Daptomicina | Daptomycin | Daptomycine | Dapto...</td>\n",
       "      <td>DOAKLVKFURWEDJ-QCMAZARJSA-N</td>\n",
       "      <td>CHEMBL4744444</td>\n",
       "      <td>CCCCCCCCCC(=O)N[C@@H](Cc1c[nH]c2ccccc12)C(=O)N...</td>\n",
       "      <td>InChI=1S/C72H101N17O26/c1-5-6-7-8-9-10-11-22-5...</td>\n",
       "      <td>False</td>\n",
       "      <td>[0.40401849150657654, 0.11877013742923737, -0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8990</th>\n",
       "      <td>DB19419</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Azadirachtin</td>\n",
       "      <td>11141-17-6</td>\n",
       "      <td>O4U1SAF85H</td>\n",
       "      <td>Azadirachtin a | Azatin | Biosal | Ecozin | Gr...</td>\n",
       "      <td>FTNJWQUOZFUQQJ-NDAWSKJSSA-N</td>\n",
       "      <td>CHEMBL509309</td>\n",
       "      <td>C/C=C(\\C)C(=O)O[C@H]1C[C@@H](OC(C)=O)[C@@]2(C(...</td>\n",
       "      <td>InChI=1S/C35H44O16/c1-8-15(2)24(38)49-18-12-19...</td>\n",
       "      <td>False</td>\n",
       "      <td>[0.6458231210708618, 0.2620610296726227, -0.11...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8991</th>\n",
       "      <td>DB19436</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L-lysyl-l-arginine</td>\n",
       "      <td>29586-66-1</td>\n",
       "      <td>HX25W9VLX3</td>\n",
       "      <td>(2s)-2-(((2s)-2,6-bis(azanyl)hexanoyl)amino)-5...</td>\n",
       "      <td>NPBGTPKLVJEOBE-IUCAKERBSA-N</td>\n",
       "      <td>CHEMBL380183</td>\n",
       "      <td>NCCCC[C@H](N)C(=O)N[C@@H](CCCN=C(N)N)C(=O)O</td>\n",
       "      <td>InChI=1S/C12H26N6O3/c13-6-2-1-4-8(14)10(19)18-...</td>\n",
       "      <td>False</td>\n",
       "      <td>[0.288204163312912, 0.22193492949008942, 0.394...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8992</th>\n",
       "      <td>DB19437</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pralurbactam</td>\n",
       "      <td>2163782-59-8</td>\n",
       "      <td>GTP46AY74R</td>\n",
       "      <td>(1R,2S,5R)-2-[[[2-[[(Aminoiminomethyl)amino]ox...</td>\n",
       "      <td>HOJIPBUGHMYVQD-RQJHMYQMSA-N</td>\n",
       "      <td>CHEMBL5314477</td>\n",
       "      <td>N=C(N)NOCCONC(=O)[C@@H]1CC[C@@H]2CN1C(=O)N2OS(...</td>\n",
       "      <td>InChI=1S/C10H18N6O8S/c11-9(12)14-23-4-3-22-13-...</td>\n",
       "      <td>False</td>\n",
       "      <td>[0.22961297631263733, -0.17735253274440765, -0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8993</th>\n",
       "      <td>DB19445</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Rocbrutinib</td>\n",
       "      <td>2485861-07-0</td>\n",
       "      <td>KD68L3GRW2</td>\n",
       "      <td>2-Propenamide, N-[5-[[6-[2-(1,3,4,6,7,8-hexahy...</td>\n",
       "      <td>OYJVFTNYBWVQHA-SANMLTNESA-N</td>\n",
       "      <td>CHEMBL5314539</td>\n",
       "      <td>C=CC(=O)Nc1cc(Nc2nc(-c3ccnc(N4CCn5c(cc6c5CC(C)...</td>\n",
       "      <td>InChI=1S/C42H51N9O5/c1-6-37(53)45-32-20-28(7-8...</td>\n",
       "      <td>False</td>\n",
       "      <td>[0.38286489248275757, -0.31730952858924866, -0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8994</th>\n",
       "      <td>DB19446</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Tulmimetostat</td>\n",
       "      <td>2567686-02-4</td>\n",
       "      <td>OUQ4N85Z0B</td>\n",
       "      <td>(2R)-7-chloro-2-[trans-4-(3-methoxyazetidin-1-...</td>\n",
       "      <td>CAAWBLRXQXMGHV-JAGNJQOISA-N</td>\n",
       "      <td>CHEMBL5095231</td>\n",
       "      <td>COC1CN([C@H]2CC[C@H]([C@@]3(C)Oc4c(Cl)cc(C(=O)...</td>\n",
       "      <td>InChI=1S/C28H36ClN3O5S/c1-15-10-23(38-5)21(27(...</td>\n",
       "      <td>False</td>\n",
       "      <td>[0.5551068186759949, 0.013536312617361546, 0.0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8995 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         drug                                    accession  \\\n",
       "0     DB00014                         BTD00113 | BIOD00113   \n",
       "1     DB00027                         BTD00036 | BIOD00036   \n",
       "2     DB00035  BTD00112 | BTD00061 | BIOD00112 | BIOD00061   \n",
       "3     DB00050             BTD00115 | APRD00686 | BIOD00115   \n",
       "4     DB00080                         BTD00111 | BIOD00111   \n",
       "...       ...                                          ...   \n",
       "8990  DB19419                                          NaN   \n",
       "8991  DB19436                                          NaN   \n",
       "8992  DB19437                                          NaN   \n",
       "8993  DB19445                                          NaN   \n",
       "8994  DB19446                                          NaN   \n",
       "\n",
       "               drug_name           cas        unii  \\\n",
       "0              Goserelin    65807-02-5  0F65R8P09N   \n",
       "1           Gramicidin D     1405-97-6  5IE62321P4   \n",
       "2           Desmopressin    16679-58-6  ENR1LLB0FP   \n",
       "3             Cetrorelix   120287-85-6  OON1HFZ4BA   \n",
       "4             Daptomycin   103060-53-3  NWQ5N31VKK   \n",
       "...                  ...           ...         ...   \n",
       "8990        Azadirachtin    11141-17-6  O4U1SAF85H   \n",
       "8991  L-lysyl-l-arginine    29586-66-1  HX25W9VLX3   \n",
       "8992        Pralurbactam  2163782-59-8  GTP46AY74R   \n",
       "8993         Rocbrutinib  2485861-07-0  KD68L3GRW2   \n",
       "8994       Tulmimetostat  2567686-02-4  OUQ4N85Z0B   \n",
       "\n",
       "                                                synonym  \\\n",
       "0                                Goserelin | Goserelina   \n",
       "1     Bacillus brevis gramicidin D | Gramicidin | Gr...   \n",
       "2     1-(3-mercaptopropionic acid)-8-D-arginine-vaso...   \n",
       "3                             Cetrorelix | Cetrorelixum   \n",
       "4     Daptomicina | Daptomycin | Daptomycine | Dapto...   \n",
       "...                                                 ...   \n",
       "8990  Azadirachtin a | Azatin | Biosal | Ecozin | Gr...   \n",
       "8991  (2s)-2-(((2s)-2,6-bis(azanyl)hexanoyl)amino)-5...   \n",
       "8992  (1R,2S,5R)-2-[[[2-[[(Aminoiminomethyl)amino]ox...   \n",
       "8993  2-Propenamide, N-[5-[[6-[2-(1,3,4,6,7,8-hexahy...   \n",
       "8994  (2R)-7-chloro-2-[trans-4-(3-methoxyazetidin-1-...   \n",
       "\n",
       "                              key         chembl  \\\n",
       "0     BLCLNMBMMGCOAS-URPVMXJPSA-N  CHEMBL1201247   \n",
       "1     NDAYQJDHGXTBJL-MWWSRJDJSA-N   CHEMBL557217   \n",
       "2     NFLWUMRGJYTJIN-PNIOQBSNSA-N     CHEMBL1429   \n",
       "3     SBNPWPIBESPSIF-MHWMIDJBSA-N  CHEMBL1200490   \n",
       "4     DOAKLVKFURWEDJ-QCMAZARJSA-N  CHEMBL4744444   \n",
       "...                           ...            ...   \n",
       "8990  FTNJWQUOZFUQQJ-NDAWSKJSSA-N   CHEMBL509309   \n",
       "8991  NPBGTPKLVJEOBE-IUCAKERBSA-N   CHEMBL380183   \n",
       "8992  HOJIPBUGHMYVQD-RQJHMYQMSA-N  CHEMBL5314477   \n",
       "8993  OYJVFTNYBWVQHA-SANMLTNESA-N  CHEMBL5314539   \n",
       "8994  CAAWBLRXQXMGHV-JAGNJQOISA-N  CHEMBL5095231   \n",
       "\n",
       "                                                 smiles  \\\n",
       "0     CC(C)C[C@H](NC(=O)[C@@H](COC(C)(C)C)NC(=O)[C@H...   \n",
       "1     CC(C)C[C@@H](NC(=O)CNC(=O)[C@@H](NC=O)C(C)C)C(...   \n",
       "2     N=C(N)NCCC[C@@H](NC(=O)[C@@H]1CCCN1C(=O)[C@@H]...   \n",
       "3     CC(=O)N[C@H](Cc1ccc2ccccc2c1)C(=O)N[C@H](Cc1cc...   \n",
       "4     CCCCCCCCCC(=O)N[C@@H](Cc1c[nH]c2ccccc12)C(=O)N...   \n",
       "...                                                 ...   \n",
       "8990  C/C=C(\\C)C(=O)O[C@H]1C[C@@H](OC(C)=O)[C@@]2(C(...   \n",
       "8991        NCCCC[C@H](N)C(=O)N[C@@H](CCCN=C(N)N)C(=O)O   \n",
       "8992  N=C(N)NOCCONC(=O)[C@@H]1CC[C@@H]2CN1C(=O)N2OS(...   \n",
       "8993  C=CC(=O)Nc1cc(Nc2nc(-c3ccnc(N4CCn5c(cc6c5CC(C)...   \n",
       "8994  COC1CN([C@H]2CC[C@H]([C@@]3(C)Oc4c(Cl)cc(C(=O)...   \n",
       "\n",
       "                                                  inchi  generated  \\\n",
       "0     InChI=1S/C59H84N18O14/c1-31(2)22-40(49(82)68-3...      False   \n",
       "1     InChI=1S/C96H135N19O16/c1-50(2)36-71(105-79(11...      False   \n",
       "2     InChI=1S/C46H64N14O12S2/c47-35(62)15-14-29-40(...      False   \n",
       "3     InChI=1S/C70H92ClN17O14/c1-39(2)31-52(61(94)82...      False   \n",
       "4     InChI=1S/C72H101N17O26/c1-5-6-7-8-9-10-11-22-5...      False   \n",
       "...                                                 ...        ...   \n",
       "8990  InChI=1S/C35H44O16/c1-8-15(2)24(38)49-18-12-19...      False   \n",
       "8991  InChI=1S/C12H26N6O3/c13-6-2-1-4-8(14)10(19)18-...      False   \n",
       "8992  InChI=1S/C10H18N6O8S/c11-9(12)14-23-4-3-22-13-...      False   \n",
       "8993  InChI=1S/C42H51N9O5/c1-6-37(53)45-32-20-28(7-8...      False   \n",
       "8994  InChI=1S/C28H36ClN3O5S/c1-15-10-23(38-5)21(27(...      False   \n",
       "\n",
       "                                              embedding  \n",
       "0     [0.26175203919410706, 0.019564686343073845, -0...  \n",
       "1     [0.505556583404541, 0.0775429829955101, -0.058...  \n",
       "2     [0.1573418825864792, -0.010056326165795326, -0...  \n",
       "3     [0.24220487475395203, 0.14190606772899628, -0....  \n",
       "4     [0.40401849150657654, 0.11877013742923737, -0....  \n",
       "...                                                 ...  \n",
       "8990  [0.6458231210708618, 0.2620610296726227, -0.11...  \n",
       "8991  [0.288204163312912, 0.22193492949008942, 0.394...  \n",
       "8992  [0.22961297631263733, -0.17735253274440765, -0...  \n",
       "8993  [0.38286489248275757, -0.31730952858924866, -0...  \n",
       "8994  [0.5551068186759949, 0.013536312617361546, 0.0...  \n",
       "\n",
       "[8995 rows x 12 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# DrugBank\n",
    "\n",
    "DrugBank = pd.read_csv(\n",
    "    \"./data/drugbank_all_drugbank_vocabulary.csv.zip\",\n",
    "    compression=\"zip\",\n",
    "    sep=\",\",\n",
    "    header=0,\n",
    "    names=[\"drug\", \"accession\", \"drug_name\", \"cas\", \"unii\", \"synonym\", \"key\"],\n",
    "    dtype={\"smiles\": str}\n",
    ")\n",
    "\n",
    "ChemRepresentation = pd.read_csv(\n",
    "    \"./data/chembl_35_chemreps.txt.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    header=0,\n",
    "    names=[\"chembl\", \"smiles\", \"inchi\", \"key\"],\n",
    ")\n",
    "\n",
    "Drug = DrugBank.merge(ChemRepresentation, on=\"key\", how=\"inner\")\n",
    "Drug = Drug.dropna(subset=[\"smiles\"])\n",
    "Drug[\"smiles\"] = Drug[\"smiles\"].astype(str).str.strip()\n",
    "Drug[\"generated\"] = False\n",
    "\n",
    "embeddings = []\n",
    "for index, row in Drug.iterrows():\n",
    "    smiles = row[\"smiles\"]\n",
    "    embedding = get_chemberta_embedding(smiles)\n",
    "    embeddings.append(embedding)\n",
    "\n",
    "Drug[\"embedding\"] = embeddings\n",
    "\n",
    "Drug\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PDB</th>\n",
       "      <th>CHAIN</th>\n",
       "      <th>SP_PRIMARY</th>\n",
       "      <th>GENE_ID</th>\n",
       "      <th>TRANSCRIPT_ID</th>\n",
       "      <th>TRANSLATION_ID</th>\n",
       "      <th>EXON_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>101m</td>\n",
       "      <td>A</td>\n",
       "      <td>P02185</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_001290722.1-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>101m</td>\n",
       "      <td>A</td>\n",
       "      <td>P02185</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_001290722.1-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>101m</td>\n",
       "      <td>A</td>\n",
       "      <td>P02185</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_001290722.1-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>102l</td>\n",
       "      <td>A</td>\n",
       "      <td>P00720</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAD42568-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>102m</td>\n",
       "      <td>A</td>\n",
       "      <td>P02185</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_001290722.1-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    PDB CHAIN SP_PRIMARY GENE_ID TRANSCRIPT_ID TRANSLATION_ID  \\\n",
       "0  101m     A     P02185     NaN           NaN            NaN   \n",
       "1  101m     A     P02185     NaN           NaN            NaN   \n",
       "2  101m     A     P02185     NaN           NaN            NaN   \n",
       "3  102l     A     P00720     NaN           NaN            NaN   \n",
       "4  102m     A     P02185     NaN           NaN            NaN   \n",
       "\n",
       "            EXON_ID  \n",
       "0  NM_001290722.1-1  \n",
       "1  NM_001290722.1-2  \n",
       "2  NM_001290722.1-3  \n",
       "3        AAD42568-1  \n",
       "4  NM_001290722.1-1  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PDB = pd.read_csv(\n",
    "    \"./data/pdb_chain_ensembl.csv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\",\",\n",
    "    header=1,\n",
    "    # names=[\"ensg\", \"ensp\"],\n",
    ")\n",
    "\n",
    "PDB.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Interactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug</th>\n",
       "      <th>drug_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DB00862</td>\n",
       "      <td>DB00966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DB00575</td>\n",
       "      <td>DB00806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DB01242</td>\n",
       "      <td>DB08893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DB01151</td>\n",
       "      <td>DB08883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DB01235</td>\n",
       "      <td>DB01275</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      drug drug_target\n",
       "0  DB00862     DB00966\n",
       "1  DB00575     DB00806\n",
       "2  DB01242     DB08893\n",
       "3  DB01151     DB08883\n",
       "4  DB01235     DB01275"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drug-Drug interaction\n",
    "\n",
    "ChCh = pd.read_csv(\n",
    "    \"./data/ChCh-Miner_durgbank-chem-chem.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    names=[\"drug\", \"drug_target\"],\n",
    ")\n",
    "\n",
    "ChCh.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This code merges drug-gene interaction data (ChG) with gene information (Gene) using the uniprot_ids "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DB00357</td>\n",
       "      <td>CYP11A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DB02721</td>\n",
       "      <td>ADH1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DB00773</td>\n",
       "      <td>PTGS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DB07138</td>\n",
       "      <td>MAPK14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DB08136</td>\n",
       "      <td>CDK2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      drug     gene\n",
       "0  DB00357  CYP11A1\n",
       "1  DB02721    ADH1B\n",
       "2  DB00773    PTGS1\n",
       "3  DB07138   MAPK14\n",
       "4  DB08136     CDK2"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drug-Gene interaction\n",
    "\n",
    "ChG = pd.read_csv(\n",
    "    \"./data/ChG-Miner_miner-chem-gene.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    header=0,\n",
    "    names=[\"drug\", \"uniprot_ids\"],\n",
    ")\n",
    "\n",
    "ChG = ChG.merge(Gene[['uniprot_ids', 'gene']], on='uniprot_ids', how='left')\n",
    "ChG.drop(columns=['uniprot_ids'], inplace=True)\n",
    "ChG.dropna(inplace=True)\n",
    "\n",
    "ChG.head(5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mesh</th>\n",
       "      <th>drug</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MESH:D005923</td>\n",
       "      <td>DB00564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MESH:D009503</td>\n",
       "      <td>DB01072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MESH:D016115</td>\n",
       "      <td>DB01759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MESH:D018476</td>\n",
       "      <td>DB00451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MESH:C567059</td>\n",
       "      <td>DB00641</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           mesh     drug\n",
       "0  MESH:D005923  DB00564\n",
       "1  MESH:D009503  DB01072\n",
       "2  MESH:D016115  DB01759\n",
       "3  MESH:D018476  DB00451\n",
       "4  MESH:C567059  DB00641"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Disease-Drug interaction\n",
    "\n",
    "DCh = pd.read_csv(\n",
    "    \"./data/DCh-Miner_miner-disease-chemical.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    header=0,\n",
    "    names=[\"mesh\", \"drug\"],\n",
    ")\n",
    "\n",
    "DCh.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>doid</th>\n",
       "      <th>doid_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DOID:0001816</td>\n",
       "      <td>DOID:1115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DOID:0002116</td>\n",
       "      <td>DOID:10124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DOID:0014667</td>\n",
       "      <td>DOID:4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DOID:0050004</td>\n",
       "      <td>DOID:10400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DOID:0050012</td>\n",
       "      <td>DOID:934</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           doid doid_target\n",
       "0  DOID:0001816   DOID:1115\n",
       "1  DOID:0002116  DOID:10124\n",
       "2  DOID:0014667      DOID:4\n",
       "3  DOID:0050004  DOID:10400\n",
       "4  DOID:0050012    DOID:934"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Disease-Disease interaction\n",
    "\n",
    "DD = pd.read_csv(\n",
    "    \"./data/DD-Miner_miner-disease-disease.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    header=0,\n",
    "    names=[\"doid\", \"doid_target\"],\n",
    ")\n",
    "\n",
    "DD.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mesh</th>\n",
       "      <th>go</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MESH:D000037</td>\n",
       "      <td>GO:0009257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MESH:C536409</td>\n",
       "      <td>GO:0009257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MESH:D009436</td>\n",
       "      <td>GO:0009257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MESH:D000860</td>\n",
       "      <td>GO:0009258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MESH:D008106</td>\n",
       "      <td>GO:0009258</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           mesh          go\n",
       "0  MESH:D000037  GO:0009257\n",
       "1  MESH:C536409  GO:0009257\n",
       "2  MESH:D009436  GO:0009257\n",
       "3  MESH:D000860  GO:0009258\n",
       "4  MESH:D008106  GO:0009258"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Disease-Function interaction\n",
    "\n",
    "DF = pd.read_csv(\n",
    "    \"./data/DF-Miner_miner-disease-function.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    header=0,\n",
    "    names=[\"mesh\", \"go\"],\n",
    ")\n",
    "\n",
    "DF.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This code merges disease-gene interaction data (DG) with gene information (Gene) using the uniprot_ids "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mesh</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MESH:D055370</td>\n",
       "      <td>PSG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MESH:D007410</td>\n",
       "      <td>KHSRP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MESH:D014062</td>\n",
       "      <td>LAIR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MESH:D054549</td>\n",
       "      <td>TRIB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MESH:D009771</td>\n",
       "      <td>CEP152</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           mesh    gene\n",
       "1  MESH:D055370    PSG1\n",
       "2  MESH:D007410   KHSRP\n",
       "3  MESH:D014062   LAIR2\n",
       "4  MESH:D054549   TRIB1\n",
       "5  MESH:D009771  CEP152"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Disease-Gene interaction\n",
    "\n",
    "DG = pd.read_csv(\n",
    "    \"./data/DG-Miner_miner-disease-gene.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    header=0,\n",
    "    names=[\"mesh\", \"uniprot_ids\"],\n",
    ")\n",
    "\n",
    "DG = DG.merge(Gene[['uniprot_ids', 'gene']], on='uniprot_ids', how='left')\n",
    "DG.drop(columns=['uniprot_ids'], inplace=True)\n",
    "DG.dropna(inplace=True)\n",
    "\n",
    "DG.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>go</th>\n",
       "      <th>go_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0008296</td>\n",
       "      <td>GO:0008408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0016811</td>\n",
       "      <td>GO:0033970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0045222</td>\n",
       "      <td>GO:0045223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0021803</td>\n",
       "      <td>GO:0030031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0033574</td>\n",
       "      <td>GO:1901654</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           go   go_target\n",
       "0  GO:0008296  GO:0008408\n",
       "1  GO:0016811  GO:0033970\n",
       "2  GO:0045222  GO:0045223\n",
       "3  GO:0021803  GO:0030031\n",
       "4  GO:0033574  GO:1901654"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Function-Function interaction\n",
    "\n",
    "FF = pd.read_csv(\n",
    "    \"./data/FF-Miner_miner-func-func.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    header=0,\n",
    "    names=[\"go\", \"go_target\"],\n",
    ")\n",
    "\n",
    "FF.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>go</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0005509</td>\n",
       "      <td>PDCD6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0004672</td>\n",
       "      <td>CDK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0005524</td>\n",
       "      <td>CDK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0005634</td>\n",
       "      <td>CDK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0005737</td>\n",
       "      <td>CDK1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           go   gene\n",
       "0  GO:0005509  PDCD6\n",
       "1  GO:0004672   CDK1\n",
       "2  GO:0005524   CDK1\n",
       "3  GO:0005634   CDK1\n",
       "4  GO:0005737   CDK1"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Gene-Function interaction (ONLY HUMANS)\n",
    "\n",
    "\n",
    "GF = pd.read_csv(\n",
    "    \"./data/GF-Miner_miner-gene-function.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    skipinitialspace=True\n",
    ")\n",
    "\n",
    "GF = GF.rename(\n",
    "    columns={\n",
    "        \"# GO_ID\": \"go\",\n",
    "        \"Gene\": \"gene\",\n",
    "        \"C8\": \"go_category\",\n",
    "        \"C10\": \"protein\",\n",
    "        \"C12\": \"organism\",\n",
    "        \"C13\": \"date\"\n",
    "    }\n",
    ")\n",
    "\n",
    "GF = GF[GF[\"organism\"] == \"taxon:9606\"]\n",
    "GF = GF[[\"go\", \"gene\"]]\n",
    "\n",
    "GF.head(5)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This code processes gene-protein relationships by merging gene symbols (gene) and PDB structural data (pdb) using ensp id's. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>pdb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MT-ND1</td>\n",
       "      <td>5xtc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MT-ND1</td>\n",
       "      <td>5xtd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MT-ND2</td>\n",
       "      <td>5xtc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MT-ND2</td>\n",
       "      <td>5xtd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MT-ND2</td>\n",
       "      <td>5xth</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene   pdb\n",
       "0  MT-ND1  5xtc\n",
       "1  MT-ND1  5xtd\n",
       "2  MT-ND2  5xtc\n",
       "3  MT-ND2  5xtd\n",
       "4  MT-ND2  5xth"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Gene-Protien interaction\n",
    "\n",
    "GP = pd.read_csv(\n",
    "    \"./data/GP-Miner_miner-gene-protein.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    header=0,\n",
    "    names=[\"ensg\", \"ensp\"],\n",
    ")\n",
    "\n",
    "ENSP = GP[['ensp']].copy()\n",
    "\n",
    "GP = GP.merge(Gene[['ensg', 'gene']], on='ensg', how='left')\n",
    "GP = GP.merge(PDB, left_on=\"ensp\", right_on=\"TRANSLATION_ID\", how=\"left\")\n",
    "GP = GP[[\"gene\", \"PDB\"]].rename(columns={\"PDB\": \"pdb\"})\n",
    "\n",
    "GP.dropna(inplace=True)\n",
    "GP.drop_duplicates(inplace=True)\n",
    "\n",
    "GP.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This code merges the ENSP IDs with PDB structural data, keeping only matching entries. It extracts and renames the PDB column to pdb, removes duplicate pdb values, and retains only the pdb column in the final DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pdb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5xtc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5xtd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5xth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5xti</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>5z62</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    pdb\n",
       "0  5xtc\n",
       "1  5xtd\n",
       "4  5xth\n",
       "5  5xti\n",
       "7  5z62"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Protein = ENSP.merge(PDB, left_on=\"ensp\", right_on=\"TRANSLATION_ID\", how=\"inner\") \n",
    "\n",
    "Protein = Protein[[\"ensp\", \"PDB\"]].rename(columns={\"PDB\": \"pdb\"})\n",
    "Protein.drop_duplicates(subset=[\"pdb\"], inplace=True)\n",
    "\n",
    "Protein = Protein[[\"pdb\"]]\n",
    "\n",
    "Protein.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pushing to DB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Database 'NeuThera' already exists.\n"
     ]
    }
   ],
   "source": [
    "sys_db = ArangoClient(hosts=\"http://localhost:8529\").db('_system', username='root', password='openSesame')\n",
    "\n",
    "# Check if the database exists\n",
    "if not sys_db.has_database('NeuThera'):\n",
    "    sys_db.create_database('NeuThera')\n",
    "    print(\"Database 'NeuThera' created successfully!\")\n",
    "else:\n",
    "    print(\"Database 'NeuThera' already exists.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "db = ArangoClient(hosts=\"http://localhost:8529\").db('NeuThera', username='root', password='openSesame')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This code ensures that the necessary ArangoDB collections exist and then inserts nodes from multiple DataFrames (Drug, Gene, Doid, Mesh, Omim, Protein). It processes data in batches, converting rows into node documents with a unique _key, and handles errors during insertion."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adding Nodes\n",
    "\n",
    "collections = [\"drug\", \"gene\", \"disease\", \"protein\"]\n",
    "\n",
    "for col in collections:\n",
    "    if not db.has_collection(col):\n",
    "        db.create_collection(col)\n",
    "\n",
    "def add_nodes(df, label, key_column, batch_size=100000):\n",
    "    batch = []\n",
    "    \n",
    "    for _, row in df.iterrows():\n",
    "        node_key = str(row[key_column]) if pd.notna(row[key_column]) else \"NotAvailable\"\n",
    "        node_id = f\"{label}/{node_key}\"\n",
    "        attributes = row.fillna(\"NaN\").drop(key_column).to_dict()\n",
    "        \n",
    "        node = {\"_key\": node_key, **attributes}\n",
    "        batch.append(node)\n",
    "        \n",
    "        if len(batch) >= batch_size:\n",
    "            try:\n",
    "                db[label].insert_many(batch, overwrite=True)\n",
    "                print(f\"Inserted {len(batch)} nodes into {label}\")\n",
    "                batch.clear()\n",
    "            except Exception as e:\n",
    "                print(f\"Error inserting batch: {e}\")\n",
    "    \n",
    "    if batch:\n",
    "        try:\n",
    "            db[label].insert_many(batch, overwrite=True)\n",
    "            print(f\"Inserted {len(batch)} nodes into {label}\")\n",
    "        except Exception as e:\n",
    "            print(f\"Error inserting final batch: {e}\")\n",
    "    \n",
    "\n",
    "add_nodes(Drug, \"drug\", \"drug\")\n",
    "add_nodes(Gene, \"gene\", \"gene\")\n",
    "add_nodes(Doid, \"disease\", \"doid\")\n",
    "add_nodes(Mesh, \"disease\", \"mesh\")\n",
    "add_nodes(Omim, \"disease\", \"omim\")\n",
    "add_nodes(Protein, \"protein\", \"pdb\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This code creates edge collections in ArangoDB ,if they don’t exist and inserts edges from various DataFrames. It processes data in batches, linking source and destination nodes while storing additional attributes\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "edge_collections = [\"drug-drug\", \"drug-gene\", \"drug-protein\", \"disease-drug\", \"disease-disease\", \"disease-function\", \"disease-gene\", \"function-function\", \"gene-function\", \"gene-protein\"]\n",
    "\n",
    "for edge_col in edge_collections:\n",
    "    if not db.has_collection(edge_col):\n",
    "        db.create_collection(edge_col, edge=True)\n",
    "\n",
    "def add_edges(df, src_label, dst_label, src_col, dst_col, edge_collection, batch_size=100000):\n",
    "    batch = []\n",
    "    \n",
    "    for _, row in df.iterrows():\n",
    "        src_node = f\"{src_label}/{str(row[src_col])}\"\n",
    "        dst_node = f\"{dst_label}/{str(row[dst_col])}\"\n",
    "\n",
    "        attributes = row.fillna(\"NaN\").drop([src_col, dst_col]).to_dict()\n",
    "        \n",
    "        edge = {\n",
    "            '_from': src_node,\n",
    "            '_to': dst_node,\n",
    "            **attributes\n",
    "        }\n",
    "        \n",
    "        batch.append(edge)\n",
    "        \n",
    "        if len(batch) >= batch_size:\n",
    "            try:\n",
    "                db[edge_collection].insert_many(batch, overwrite=True)\n",
    "                print(f\"Inserted {len(batch)} edges into {edge_collection}\")\n",
    "                batch.clear()\n",
    "            except Exception as e:\n",
    "                print(f\"Error inserting batch: {e}\")\n",
    "    \n",
    "    if batch:\n",
    "        try:\n",
    "            db[edge_collection].insert_many(batch, overwrite=True)\n",
    "            print(f\"Inserted {len(batch)} edges into {edge_collection}\")\n",
    "        except Exception as e:\n",
    "            print(f\"Error inserting final batch: {e}\")\n",
    "\n",
    "add_edges(ChCh, \"drug\", \"drug\", \"drug\", \"drug_target\", \"drug-drug\")\n",
    "add_edges(ChG, \"drug\", \"gene\", \"drug\", \"gene\", \"drug-gene\")\n",
    "add_edges(DCh, \"disease\", \"drug\", \"mesh\", \"drug\", \"disease-drug\")\n",
    "add_edges(DD, \"disease\", \"disease\", \"doid\", \"doid_target\", \"disease-disease\")\n",
    "add_edges(DF, \"disease\", \"go\", \"mesh\", \"go\", \"disease-function\")\n",
    "add_edges(DG, \"disease\", \"gene\", \"mesh\", \"gene\", \"disease-gene\")\n",
    "add_edges(FF, \"go\", \"go\", \"go\", \"go_target\", \"function-function\")\n",
    "add_edges(GF, \"go\", \"gene\", \"go\", \"gene\", \"gene-function\")\n",
    "add_edges(GP, \"gene\", \"protein\", \"gene\", \"pdb\", \"gene-protein\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This code initializes the NeuThera graph in ArangoDB, defining relationships between nodes by setting up edge definitions. It ensures all specified edges correctly link their respective node collections, creating a structured biomedical knowledge graph.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NeuThera graph successfully created and linked with node and edge collections!\n"
     ]
    }
   ],
   "source": [
    "if not db.has_graph(\"NeuThera\"):\n",
    "    db.create_graph(\"NeuThera\")\n",
    "\n",
    "neuthera_graph = db.graph(\"NeuThera\")\n",
    "\n",
    "edge_definitions = [\n",
    "    (\"drug-drug\", \"drug\", \"drug\"),\n",
    "    (\"drug-gene\", \"drug\", \"gene\"),\n",
    "    (\"disease-drug\", \"disease\", \"drug\"),\n",
    "    (\"disease-disease\", \"disease\", \"disease\"),\n",
    "    (\"disease-function\", \"disease\", \"go\"),\n",
    "    (\"disease-gene\", \"disease\", \"gene\"),\n",
    "    (\"function-function\", \"go\", \"go\"),\n",
    "    (\"gene-function\", \"gene\", \"go\"),\n",
    "    (\"gene-protein\", \"gene\", \"protein\"),\n",
    "    (\"drug-protein\", \"drug\", \"protein\")\n",
    "]\n",
    "\n",
    "for edge_col, from_col, to_col in edge_definitions:\n",
    "    if edge_col not in neuthera_graph.edge_definitions():\n",
    "        neuthera_graph.create_edge_definition(\n",
    "            edge_collection=edge_col,\n",
    "            from_vertex_collections=[from_col],\n",
    "            to_vertex_collections=[to_col]\n",
    "        )\n",
    "\n",
    "print(\"NeuThera graph successfully created and linked with node and edge collections!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Modelling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "load_dotenv()\n",
    "openai_api_key = os.getenv(\"OPENAI_API_KEY\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "arango_graph = ArangoGraph(db)\n",
    "\n",
    "drug_collection = db.collection('drug')\n",
    "link_collection = db.collection('drug-protein') "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This code retrieves drug embeddings from ArangoDB using an AQL query, storing the _key values in drug_keys and their corresponding embeddings in embeddings. It then converts the embeddings into a NumPy array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Embeddings shape: (9014, 768)\n",
      "Number of compounds: 9014\n"
     ]
    }
   ],
   "source": [
    "cursor = db.aql.execute(\"FOR doc IN drug RETURN {key: doc._key, embedding: doc.embedding}\")\n",
    "\n",
    "drug_keys = []\n",
    "embeddings = []\n",
    "\n",
    "for doc in cursor:\n",
    "    if doc and \"embedding\" in doc and \"key\" in doc:\n",
    "        drug_keys.append(doc[\"key\"])\n",
    "        embeddings.append(doc[\"embedding\"])\n",
    "\n",
    "embeddings = np.array(embeddings, dtype=np.float32)\n",
    "\n",
    "print(\"Embeddings shape:\", embeddings.shape)\n",
    "print(\"Number of compounds:\", len(drug_keys))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tooling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Starting Tool Kit :- \n",
    "\n",
    "1. text_to_aql :- Executes a Natural Language Query in ArangoDB, and returns the result as text.\n",
    "2. predict_binding_affinity :- Predicts the binding affinity for given drug and target sequences.\n",
    "3. get_amino_acid_sequence_from_pdb :-  Extracts amino acid sequences from a given PDB structure file in CIF format\n",
    "4. get_chemberta_embedding :- Generate a ChemBERTa vector embedding for a given molecule represented as a SMILES string.\n",
    "5. prepare_pdb_data :- Checks if the PDB data for the given PDB ID is available. If not, downloads and processes the data.\n",
    "6. generate_compounds :- Generates and sorts compounds based on similarity to a reference molecule, all generated compounds are added back to the database    for futher inference.\n",
    "7. generate_report :- Generate a report in CSV format with a timestamped filename. This function uses pandas to create a CSV.\n",
    "8. find_similar_drugs :- Finds the top K most similar drugs based on given smile of a query molecule. Automatically gets vector embeddings.\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tooling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "@tool\n",
    "def text_to_aql(query: str):\n",
    "    \"\"\"Execute a Natural Language Query in ArangoDB, and return the result as text.\"\"\"\n",
    "    \n",
    "    llm = ChatOpenAI(temperature=0, model_name=\"gpt-4o\")\n",
    "\n",
    "    chain = ArangoGraphQAChain.from_llm(\n",
    "        llm=llm,\n",
    "        graph=arango_graph,  # Assuming arango_graph is already initialized\n",
    "        verbose=True,\n",
    "        allow_dangerous_requests=True\n",
    "    )\n",
    "    \n",
    "    result = chain.invoke(query)\n",
    "\n",
    "    return str(result[\"result\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "@tool\n",
    "def predict_binding_affinity(X_drug, X_target, y=[7.635]):\n",
    "    \"\"\"\n",
    "    Predicts the binding affinity for given drug and target sequences.\n",
    "\n",
    "    Parameters:\n",
    "    X_drug (list): List containing the SMILES representation of the drug.\n",
    "    X_target (list): List containing the amino acid sequence of the protein target.\n",
    "\n",
    "    Returns:\n",
    "    float: Predicted binding affinity (log(Kd) or log(Ki)).\n",
    "    \"\"\"\n",
    "\n",
    "    print(\"Predicting binding affinity: \", X_drug, X_target)\n",
    "    \n",
    "    model = models.model_pretrained(path_dir='DTI_model')\n",
    "\n",
    "    X_pred = utils.data_process(X_drug, X_target, y,\n",
    "                                drug_encoding='CNN', \n",
    "                                target_encoding='CNN', \n",
    "                                split_method='no_split')\n",
    "   \n",
    "    predictions = model.predict(X_pred)\n",
    "\n",
    "    return predictions[0]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "@tool\n",
    "def get_amino_acid_sequence_from_pdb(pdb_id):    \n",
    "    \"\"\"\n",
    "    Extracts amino acid sequences from a given PDB structure file in CIF format.\n",
    "\n",
    "    Args:\n",
    "        pdb_id (str): pdb id of the protein.\n",
    "\n",
    "    Returns:\n",
    "        dict: A dictionary where keys are chain IDs and values are amino acid sequences.\n",
    "    \"\"\"\n",
    "\n",
    "    print(\"Getting Amino Acid sequence for \", pdb_id)\n",
    "\n",
    "    cif_file_path = f\"./database/PDBlib/{pdb_id.lower()}.cif\"\n",
    "\n",
    "    parser = MMCIFParser(QUIET=True)\n",
    "    structure = parser.get_structure(\"protein\", cif_file_path)\n",
    "    \n",
    "    sequences = {}\n",
    "    for model in structure:\n",
    "        for chain in model:\n",
    "            seq = \"\".join(residue.resname for residue in chain if residue.id[0] == \" \")\n",
    "            sequences[chain.id] = seq \n",
    "            \n",
    "    return sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.append(os.path.abspath(\"./TamGen\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Helper Functions for TamGen\n",
    "\n",
    "tokenizer = AutoTokenizer.from_pretrained(\"seyonec/ChemBERTa-zinc-base-v1\")\n",
    "model = AutoModel.from_pretrained(\"seyonec/ChemBERTa-zinc-base-v1\")\n",
    "\n",
    "def get_chemberta_embedding(smiles):\n",
    "    \"\"\"\n",
    "    Generate a ChemBERTa vector embedding for a given molecule represented as a SMILES string.\n",
    "\n",
    "    Args:\n",
    "        smiles (str): A valid SMILES representation of a molecule.\n",
    "\n",
    "    Returns:\n",
    "        List[float] or None: A 768-dimensional vector as a list of floats if successful, \n",
    "                             otherwise None if the input is invalid.\n",
    "    \"\"\"\n",
    "    \n",
    "    print(\"Getting vector embedding\")\n",
    "\n",
    "    if not isinstance(smiles, str) or not smiles.strip():\n",
    "        return None \n",
    "\n",
    "    inputs = tokenizer(smiles, return_tensors=\"pt\", padding=True, truncation=True)\n",
    "    with torch.no_grad():\n",
    "        outputs = model(**inputs)\n",
    "    return outputs.last_hidden_state.mean(dim=1).tolist()[0]\n",
    "\n",
    "def generate_key(smiles):\n",
    "    \"\"\"Generate a unique _key for the compound using SMILES hash.\"\"\"\n",
    "    hash_value = hashlib.sha256(smiles.encode()).hexdigest()[:8]\n",
    "    return f\"GEN:{hash_value}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Namespace(no_progress_bar=False, log_interval=1000, log_format=None, tensorboard_logdir='', tbmf_wrapper=False, seed=1, cpu=False, fp16=False, memory_efficient_fp16=False, fp16_init_scale=128, fp16_scale_window=None, fp16_scale_tolerance=0.0, min_loss_scale=0.0001, threshold_loss_scale=None, user_dir=None, criterion='cross_entropy', tokenizer=None, bpe=None, optimizer='nag', lr_scheduler='fixed', task='translation_coord', num_workers=1, skip_invalid_size_inputs_valid_test=False, max_tokens=1024, max_sentences=None, required_batch_size_multiple=8, dataset_impl=None, gen_subset='gen_8fln', num_shards=1, shard_id=0, path='checkpoints/crossdock_pdb_A10/checkpoint_best.pt', remove_bpe=None, quiet=False, model_overrides='{}', results_path=None, beam=20, nbest=20, max_len_a=0, max_len_b=200, min_len=1, match_source_len=False, no_early_stop=False, unnormalized=False, no_beamable_mm=False, lenpen=1, unkpen=0, replace_unk=None, sacrebleu=False, score_reference=False, prefix_size=0, prefix_string=None, tag_prefix_token=False, no_repeat_ngram_size=0, sampling=False, sampling_topk=-1, sampling_topp=-1.0, temperature=1.0, diverse_beam_groups=-1, diverse_beam_strength=0.5, print_alignment=False, scorer='bleu', codegen_pov_replace_unk=True, dump_hyp=None, momentum=0.99, weight_decay=0.0, force_anneal=None, lr_shrink=0.1, warmup_updates=0, data='./TamGen_Demo_Data', source_lang='tg', target_lang='m1', lazy_load=False, raw_text=False, left_pad_source='True', left_pad_target='False', max_source_positions=1024, max_target_positions=1024, upsample_primary=1, coord_mode='raw', use_src_coord=True, use_tgt_coord=False, shuffle_input=False, vae=False, sample_beta=1.0, hint=False, hint_rate=0.5, gen_coord_noise=False, gen_rot=False, gen_vae=True, remove_prefix_fragment=False)\n",
      "| [tg] dictionary: 25 types\n",
      "| [m1] dictionary: 962 types\n",
      "| loading model(s) from checkpoints/crossdock_pdb_A10/checkpoint_best.pt\n",
      "checkpoints/crossdock_pdb_A10/checkpoint_best.pt\n",
      "gpt_model/checkpoint_best.pt\n"
     ]
    }
   ],
   "source": [
    "from TamGen_custom import TamGenCustom\n",
    "\n",
    "os.environ[\"CUDA_VISIBLE_DEVICES\"] = \"0\"\n",
    "\n",
    "worker = TamGenCustom(\n",
    "    data=\"./TamGen_Demo_Data\",\n",
    "    ckpt=\"checkpoints/crossdock_pdb_A10/checkpoint_best.pt\",\n",
    "    use_conditional=True\n",
    ")\n",
    "\n",
    "@tool\n",
    "def prepare_pdb_data(pdb_id):\n",
    "    \"\"\"\n",
    "    Checks if the PDB data for the given PDB ID is available.  \n",
    "    If not, downloads and processes the data.\n",
    "\n",
    "    ALWAYS RUN THIS FUNCTION BEFORE WORKING WITH PDB\n",
    "\n",
    "    Args:\n",
    "        pdb_id (str): PDB ID of the target structure.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    DemoDataFolder=\"TamGen_Demo_Data\"\n",
    "    ligand_inchi=None\n",
    "    thr=10\n",
    "\n",
    "    out_split = pdb_id.lower()\n",
    "    FF = glob(f\"{DemoDataFolder}/*\")\n",
    "    for ff in FF:\n",
    "        if f\"gen_{out_split}\" in ff:\n",
    "            print(f\"{pdb_id} is downloaded\")\n",
    "            return\n",
    "    \n",
    "    os.makedirs(DemoDataFolder, exist_ok=True)\n",
    "    \n",
    "    with open(\"tmp_pdb.csv\", \"w\") as fw:\n",
    "        if ligand_inchi is None:\n",
    "            print(\"pdb_id\", file=fw)\n",
    "            print(f\"{pdb_id}\", file=fw)\n",
    "        else:\n",
    "            print(\"pdb_id,ligand_inchi\", file=fw)\n",
    "            print(f\"{pdb_id},{ligand_inchi}\", file=fw)\n",
    "\n",
    "    script_path = os.path.abspath(\"TamGen/scripts/build_data/prepare_pdb_ids.py\")\n",
    "    os.system(f\"python {script_path} tmp_pdb.csv gen_{out_split} -o {DemoDataFolder} -t {thr}\")\n",
    "    os.remove(\"tmp_pdb.csv\")\n",
    "\n",
    "@tool\n",
    "def generate_compounds(pdb_id, num_samples=10, max_seed=30):\n",
    "    \"\"\"\n",
    "    Generates and sorts compounds based on similarity to a reference molecule, \n",
    "    all generated compounds are added back to the database for futher inference.\n",
    "\n",
    "    Parameters:\n",
    "    - pdb_id (str): The PDB ID of the target protein.\n",
    "    - num_samples (int): Number of compounds to generate. (DEFAULT=500)\n",
    "    - max_seed (int): Maximum seed variations. (DEFAULT=30)\n",
    "\n",
    "    Returns:\n",
    "    - dict: {\n",
    "        'generated': [list of rdkit Mol objects],\n",
    "        'reference': rdkit Mol object,\n",
    "        'reference_smile': SMILE string of the reference compound\n",
    "        'generated_smiles': [list of SMILES strings, sorted by similarity to reference]\n",
    "      }\n",
    "    \"\"\"\n",
    "\n",
    "    print(\"Generating Compounds for PDB \", pdb_id)\n",
    "    try:\n",
    "        # Ensure the required PDB data is prepared\n",
    "        # prepare_pdb_data(pdb_id)\n",
    "\n",
    "        worker.reload_data(subset=f\"gen_{pdb_id.lower()}\")\n",
    "\n",
    "        print(f\"Generating {num_samples} compounds...\")\n",
    "        generated_mols, reference_mol = worker.sample(\n",
    "            m_sample=num_samples, \n",
    "            maxseed=max_seed\n",
    "        )\n",
    "\n",
    "        if reference_mol:\n",
    "            # Ensure reference_mol is an RDKit Mol object\n",
    "            if isinstance(reference_mol, str):\n",
    "                reference_mol = Chem.MolFromSmiles(reference_mol)\n",
    "\n",
    "            fp_ref = MACCSkeys.GenMACCSKeys(reference_mol)\n",
    "\n",
    "            gens = []\n",
    "            for mol in generated_mols:\n",
    "                if isinstance(mol, str):  # Convert string SMILES to Mol\n",
    "                    mol = Chem.MolFromSmiles(mol)\n",
    "                if mol:  # Ensure conversion was successful\n",
    "                    fp = MACCSkeys.GenMACCSKeys(mol)\n",
    "                    similarity = DataStructs.FingerprintSimilarity(fp_ref, fp, metric=DataStructs.TanimotoSimilarity)\n",
    "                    gens.append((mol, similarity))\n",
    "\n",
    "            sorted_mols = [mol for mol, _ in sorted(gens, key=lambda e: e[1], reverse=True)]\n",
    "        \n",
    "        else:\n",
    "            sorted_mols = generated_mols\n",
    "\n",
    "        generated_smiles = [Chem.MolToSmiles(mol) for mol in sorted_mols if mol]\n",
    "\n",
    "        reference_smile = Chem.MolToSmiles(reference_mol)\n",
    "        \n",
    "        print(\"Inserting to ArangoDB...\")\n",
    "        for smiles in generated_smiles:\n",
    "            _key = generate_key(smiles) \n",
    "            drug_id = f\"drug/{_key}\"\n",
    "            protein_id = f\"protein/{pdb_id}\"\n",
    "\n",
    "            if drug_collection.has(_key):\n",
    "                continue\n",
    "\n",
    "            embedding = get_chemberta_embedding(smiles)\n",
    "            doc = {\n",
    "                \"_key\": _key,\n",
    "                \"_id\": drug_id, \n",
    "                \"accession\": \"NaN\",\n",
    "                \"drug_name\": \"NaN\",\n",
    "                \"cas\": \"NaN\",\n",
    "                \"unii\": \"NaN\",\n",
    "                \"synonym\": \"NaN\",\n",
    "                \"key\": \"NaN\",\n",
    "                \"chembl\": \"NaN\",\n",
    "                \"smiles\": smiles,\n",
    "                \"inchi\": \"NaN\",\n",
    "                \"generated\": True,\n",
    "                \"embedding\": embedding\n",
    "            }\n",
    "            drug_collection.insert(doc)\n",
    "\n",
    "            existing_links = list(db.aql.execute(f'''\n",
    "                FOR link IN `drug-protein` \n",
    "                FILTER link._from == \"{drug_id}\" AND link._to == \"{protein_id}\" \n",
    "                RETURN link\n",
    "            '''))\n",
    "\n",
    "            if not existing_links:\n",
    "                link_doc = {\n",
    "                    \"_from\": drug_id,\n",
    "                    \"_to\": protein_id,\n",
    "                    \"generated\": True\n",
    "                }\n",
    "                link_collection.insert(link_doc)\n",
    "\n",
    "        return {\n",
    "            \"generated\": sorted_mols,\n",
    "            \"reference\": reference_mol,\n",
    "            \"reference_smile\": reference_smile,\n",
    "            \"generated_smiles\": generated_smiles\n",
    "        }\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"Error in compound generation: {str(e)}\")\n",
    "        return {\"error\": str(e)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "@tool\n",
    "def generate_report(columns, rows):\n",
    "    \"\"\"\n",
    "    Generate a report in CSV format with a timestamped filename. This function uses pandas to create a CSV.\n",
    "    \n",
    "    Parameters:\n",
    "    columns (list): List of column names.\n",
    "    rows (list of lists): Data rows corresponding to the columns.\n",
    "    \n",
    "    Returns:\n",
    "    str: Path of the generated CSV report.\n",
    "    \"\"\"\n",
    "    timestamp = datetime.now().strftime(\"%Y%m%d_%H%M%S\")\n",
    "    filename = f\"report_{timestamp}.csv\"\n",
    "    \n",
    "    df = pd.DataFrame(rows, columns=columns)\n",
    "    df.to_csv(filename, index=False)\n",
    "    \n",
    "    return filename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "@tool\n",
    "def find_similar_drugs(smile, top_k=5):\n",
    "    \"\"\"\n",
    "    Finds the top K most similar drugs based on given smile of a query molecule. Automatically gets vector embeddings.\n",
    "\n",
    "    Args:\n",
    "        smile (string): Smile of the query molecule.\n",
    "        top_k (int, optional): Number of most similar drugs to retrieve. Default is 5.\n",
    "\n",
    "    Returns:\n",
    "        List[Dict{str, [float]}]: A list of (drug_name, similarity_score) sorted by similarity.\n",
    "    \"\"\"\n",
    "    \n",
    "    print(\"Finding similar drugs...\")\n",
    "\n",
    "    embedding = get_chemberta_embedding(smile)\n",
    "    \n",
    "    aql_query = f\"\"\"\n",
    "    LET query_vector = @query_vector\n",
    "    FOR doc IN drug\n",
    "        LET score = COSINE_SIMILARITY(doc.embedding, query_vector)\n",
    "        SORT score DESC\n",
    "        LIMIT @top_k\n",
    "        RETURN {{ drug: doc._key, similarity_score: score }}\n",
    "    \"\"\"\n",
    "    \n",
    "    cursor = db.aql.execute(aql_query, bind_vars={\"query_vector\": embedding, \"top_k\": top_k})\n",
    "    \n",
    "    return list(cursor)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Agentic RAG"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This code defines a drug discovery assistant using an LLM-powered agent (ChatOpenAI) with various tools for querying a biomedical knowledge graph. It ensures structured JSON output, avoids redundant embedding calculations, and processes user queries step-by-step using the specified tools."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "tools = [ text_to_aql, get_amino_acid_sequence_from_pdb, prepare_pdb_data, generate_compounds, predict_binding_affinity, generate_report, get_chemberta_embedding, find_similar_drugs ]\n",
    "\n",
    "def query_graph(query):\n",
    "    \n",
    "    query_template = f\"\"\"\n",
    "    USER INPUT: {query}\n",
    "\n",
    "    You are an advanced drug discovery assistant with multiple tools.\n",
    "\n",
    "    - Use your tools as needed to assist in end-to-end drug discovery and answer user queries.\n",
    "    - Always structure your output as valid JSON string so it can be parsed in python.\n",
    "    - If possible, always try to generate reports for whatever output you get. Don't generate reports for errors.\n",
    "    - Do not add explanations or any extra text.\n",
    "    - When working with multiple outputs, run functions one by one for everything unless stated otherwise by the user.\n",
    "\n",
    "    TO NOTE:\n",
    "    - DO NOT USE get_chemberta_embedding TOOL AS ALL GENERATED COMPOUNDS ALREADY HAVE THEIR EMBEDDINGS IN THE DATABASE,\n",
    "        USE AQL INSTEAD TO FETCH\n",
    "    \"\"\"\n",
    "    \n",
    "    llm = ChatOpenAI(temperature=0, model_name=\"gpt-4o\")\n",
    "    app = create_react_agent(llm, tools)    \n",
    "    final_state = app.invoke({\"messages\": [{\"role\": \"user\", \"content\": query_template}]})\n",
    "    return json.loads(final_state[\"messages\"][-1].content)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Frontend\n",
    "\n",
    "All reports are in the root query\n",
    "\n",
    "Some queries to try:\n",
    "- What proteins can you find me related to mitochondrial ribosomal protein L36? Generate some compounds and test their binding affinity\n",
    "- Take a random generated compound from the database and find top 10 most similar drugs to it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "* Running on local URL:  http://127.0.0.1:7862\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
      "To disable this warning, you can either:\n",
      "\t- Avoid using `tokenizers` before the fork if possible\n",
      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "* Running on public URL: https://99ec199764783ea872.gradio.live\n",
      "\n",
      "This share link expires in 72 hours. For free permanent hosting and GPU upgrades, run `gradio deploy` from the terminal in the working directory to deploy to Hugging Face Spaces (https://huggingface.co/spaces)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div><iframe src=\"https://99ec199764783ea872.gradio.live\" width=\"100%\" height=\"500\" allow=\"autoplay; camera; microphone; clipboard-read; clipboard-write;\" frameborder=\"0\" allowfullscreen></iframe></div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m> Entering new ArangoGraphQAChain chain...\u001b[0m\n",
      "AQL Query (1):\u001b[32;1m\u001b[1;3m\n",
      "WITH drug\n",
      "FOR compound IN drug\n",
      "FILTER compound.generated == true\n",
      "LIMIT 1\n",
      "RETURN compound.smiles\n",
      "\u001b[0m\n",
      "AQL Result:\n",
      "\u001b[32;1m\u001b[1;3m['Nc1ncnc2c1ncn2[C@@H]1O[C@H](COP(=O)(O)OP(=O)(O)OP(=O)(O)O)[C@@H](O)[C@H]1O']\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "Finding similar drugs...\n",
      "Getting vector embedding\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ArangoGraphQAChain chain...\u001b[0m\n",
      "AQL Query (1):\u001b[32;1m\u001b[1;3m\n",
      "WITH drug\n",
      "FOR compound IN drug\n",
      "RETURN compound.smiles\n",
      "\u001b[0m\n",
      "AQL Result:\n",
      "\u001b[32;1m\u001b[1;3m['CC(C)C[C@H](NC(=O)[C@@H](COC(C)(C)C)NC(=O)[C@H](Cc1ccc(O)cc1)NC(=O)[C@H](CO)NC(=O)[C@H](Cc1c[nH]c2ccccc12)NC(=O)[C@H](Cc1cnc[nH]1)NC(=O)[C@@H]1CCC(=O)N1)C(=O)N[C@@H](CCCNC(=N)N)C(=O)N1CCC[C@H]1C(=O)NNC(N)=O', 'CC(C)C[C@@H](NC(=O)CNC(=O)[C@@H](NC=O)C(C)C)C(=O)N[C@@H](C)C(=O)N[C@@H](C(=O)N[C@H](C(=O)N[C@@H](C(=O)N[C@@H](Cc1c[nH]c2ccccc12)C(=O)N[C@H](CC(C)C)C(=O)N[C@@H](Cc1c[nH]c2ccccc12)C(=O)N[C@H](CC(C)C)C(=O)N[C@@H](Cc1c[nH]c2ccccc12)C(=O)N[C@H](CC(C)C)C(=O)N[C@@H](Cc1c[nH]c2ccccc12)C(=O)NCCO)C(C)C)C(C)C)C(C)C', 'N=C(N)NCCC[C@@H](NC(=O)[C@@H]1CCCN1C(=O)[C@@H]1CSSCCC(=O)N[C@@H](Cc2ccc(O)cc2)C(=O)N[C@@H](Cc2ccccc2)C(=O)N[C@@H](CCC(N)=O)C(=O)N[C@@H](CC(N)=O)C(=O)N1)C(=O)NCC(N)=O', 'CC(=O)N[C@H](Cc1ccc2ccccc2c1)C(=O)N[C@H](Cc1ccc(Cl)cc1)C(=O)N[C@H](Cc1cccnc1)C(=O)N[C@@H](CO)C(=O)N[C@@H](Cc1ccc(O)cc1)C(=O)N[C@H](CCCNC(N)=O)C(=O)N[C@@H](CC(C)C)C(=O)N[C@@H](CCCNC(=N)N)C(=O)N1CCC[C@H]1C(=O)N[C@H](C)C(N)=O', 'CCCCCCCCCC(=O)N[C@@H](Cc1c[nH]c2ccccc12)C(=O)N[C@H](CC(N)=O)C(=O)N[C@@H](CC(=O)O)C(=O)N[C@@H]1C(=O)NCC(=O)N[C@@H](CCCN)C(=O)N[C@@H](CC(=O)O)C(=O)N[C@H](C)C(=O)N[C@@H](CC(=O)O)C(=O)NCC(=O)N[C@H](CO)C(=O)N[C@@H]([C@H](C)CC(=O)O)C(=O)N[C@@H](CC(=O)c2ccccc2N)C(=O)O[C@@H]1C', 'C/C=C/C[C@@H](C)[C@@H](O)[C@H]1C(=O)N[C@@H](CC)C(=O)N(C)CC(=O)N(C)[C@@H](CC(C)C)C(=O)N[C@@H](C(C)C)C(=O)N(C)[C@@H](CC(C)C)C(=O)N[C@@H](C)C(=O)N[C@H](C)C(=O)N(C)[C@@H](CC(C)C)C(=O)N(C)[C@@H](CC(C)C)C(=O)N(C)[C@@H](C(C)C)C(=O)N1C', 'CC(=O)N[C@H](Cc1ccc2ccccc2c1)C(=O)N[C@H](Cc1ccc(Cl)cc1)C(=O)N[C@H](Cc1cccnc1)C(=O)N[C@@H](CO)C(=O)N(C)[C@@H](Cc1ccc(O)cc1)C(=O)N[C@H](CC(N)=O)C(=O)N[C@@H](CC(C)C)C(=O)N[C@@H](CCCCNC(C)C)C(=O)N1CCC[C@H]1C(=O)N[C@H](C)C(N)=O', 'Cc1ncc(COP(=O)(O)O)c(C=O)c1O', 'Nc1nc(O)c2c(n1)NCC(CNc1ccc(C(=O)N[C@@H](CCC(=O)O)C(=O)O)cc1)N2', 'N[C@@H](Cc1c[nH]cn1)C(=O)O']\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ArangoGraphQAChain chain...\u001b[0m\n",
      "AQL Query (1):\u001b[32;1m\u001b[1;3m\n",
      "WITH drug\n",
      "FOR compound IN drug\n",
      "RETURN compound.smiles\n",
      "\u001b[0m\n",
      "AQL Result:\n",
      "\u001b[32;1m\u001b[1;3m['CC(C)C[C@H](NC(=O)[C@@H](COC(C)(C)C)NC(=O)[C@H](Cc1ccc(O)cc1)NC(=O)[C@H](CO)NC(=O)[C@H](Cc1c[nH]c2ccccc12)NC(=O)[C@H](Cc1cnc[nH]1)NC(=O)[C@@H]1CCC(=O)N1)C(=O)N[C@@H](CCCNC(=N)N)C(=O)N1CCC[C@H]1C(=O)NNC(N)=O', 'CC(C)C[C@@H](NC(=O)CNC(=O)[C@@H](NC=O)C(C)C)C(=O)N[C@@H](C)C(=O)N[C@@H](C(=O)N[C@H](C(=O)N[C@@H](C(=O)N[C@@H](Cc1c[nH]c2ccccc12)C(=O)N[C@H](CC(C)C)C(=O)N[C@@H](Cc1c[nH]c2ccccc12)C(=O)N[C@H](CC(C)C)C(=O)N[C@@H](Cc1c[nH]c2ccccc12)C(=O)N[C@H](CC(C)C)C(=O)N[C@@H](Cc1c[nH]c2ccccc12)C(=O)NCCO)C(C)C)C(C)C)C(C)C', 'N=C(N)NCCC[C@@H](NC(=O)[C@@H]1CCCN1C(=O)[C@@H]1CSSCCC(=O)N[C@@H](Cc2ccc(O)cc2)C(=O)N[C@@H](Cc2ccccc2)C(=O)N[C@@H](CCC(N)=O)C(=O)N[C@@H](CC(N)=O)C(=O)N1)C(=O)NCC(N)=O', 'CC(=O)N[C@H](Cc1ccc2ccccc2c1)C(=O)N[C@H](Cc1ccc(Cl)cc1)C(=O)N[C@H](Cc1cccnc1)C(=O)N[C@@H](CO)C(=O)N[C@@H](Cc1ccc(O)cc1)C(=O)N[C@H](CCCNC(N)=O)C(=O)N[C@@H](CC(C)C)C(=O)N[C@@H](CCCNC(=N)N)C(=O)N1CCC[C@H]1C(=O)N[C@H](C)C(N)=O', 'CCCCCCCCCC(=O)N[C@@H](Cc1c[nH]c2ccccc12)C(=O)N[C@H](CC(N)=O)C(=O)N[C@@H](CC(=O)O)C(=O)N[C@@H]1C(=O)NCC(=O)N[C@@H](CCCN)C(=O)N[C@@H](CC(=O)O)C(=O)N[C@H](C)C(=O)N[C@@H](CC(=O)O)C(=O)NCC(=O)N[C@H](CO)C(=O)N[C@@H]([C@H](C)CC(=O)O)C(=O)N[C@@H](CC(=O)c2ccccc2N)C(=O)O[C@@H]1C', 'C/C=C/C[C@@H](C)[C@@H](O)[C@H]1C(=O)N[C@@H](CC)C(=O)N(C)CC(=O)N(C)[C@@H](CC(C)C)C(=O)N[C@@H](C(C)C)C(=O)N(C)[C@@H](CC(C)C)C(=O)N[C@@H](C)C(=O)N[C@H](C)C(=O)N(C)[C@@H](CC(C)C)C(=O)N(C)[C@@H](CC(C)C)C(=O)N(C)[C@@H](C(C)C)C(=O)N1C', 'CC(=O)N[C@H](Cc1ccc2ccccc2c1)C(=O)N[C@H](Cc1ccc(Cl)cc1)C(=O)N[C@H](Cc1cccnc1)C(=O)N[C@@H](CO)C(=O)N(C)[C@@H](Cc1ccc(O)cc1)C(=O)N[C@H](CC(N)=O)C(=O)N[C@@H](CC(C)C)C(=O)N[C@@H](CCCCNC(C)C)C(=O)N1CCC[C@H]1C(=O)N[C@H](C)C(N)=O', 'Cc1ncc(COP(=O)(O)O)c(C=O)c1O', 'Nc1nc(O)c2c(n1)NCC(CNc1ccc(C(=O)N[C@@H](CCC(=O)O)C(=O)O)cc1)N2', 'N[C@@H](Cc1c[nH]cn1)C(=O)O']\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new ArangoGraphQAChain chain...\u001b[0m\n",
      "AQL Query (1):\u001b[32;1m\u001b[1;3m\n",
      "WITH drug\n",
      "FOR compound IN drug\n",
      "LIMIT 1\n",
      "RETURN compound.smiles\n",
      "\u001b[0m\n",
      "AQL Result:\n",
      "\u001b[32;1m\u001b[1;3m['CC(C)C[C@H](NC(=O)[C@@H](COC(C)(C)C)NC(=O)[C@H](Cc1ccc(O)cc1)NC(=O)[C@H](CO)NC(=O)[C@H](Cc1c[nH]c2ccccc12)NC(=O)[C@H](Cc1cnc[nH]1)NC(=O)[C@@H]1CCC(=O)N1)C(=O)N[C@@H](CCCNC(=N)N)C(=O)N1CCC[C@H]1C(=O)NNC(N)=O']\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "Finding similar drugs...\n",
      "Getting vector embedding\n"
     ]
    }
   ],
   "source": [
    "import gradio as gr\n",
    "\n",
    "gr.Interface(fn=query_graph, inputs=\"text\", outputs=\"text\").launch(share=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "TamGen",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
