{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pre-Processing dataset\n",
    "\n",
    "We create a KG in NetworkX-arangoDB\n",
    "\n",
    "This requires downloading multi-modality datasets from [Bio Snap](https://snap.stanford.edu/biodata/)\n",
    "\n",
    "##### Modalities:\n",
    "- Drug-Drug interaction with side-effects\n",
    "- Drug-Gene interaction\n",
    "- Disease-Drug interaction\n",
    "- Disease-Disease interaction\n",
    "- Disease-Function interaction\n",
    "- Function-Function interaction\n",
    "- Gene-Function interaction\n",
    "- Gene-Gene interaction\n",
    "- Gene-Protien interaction\n",
    "- Genomic Region-Genomic Region interaction\n",
    "- Protien-Protien interaction\n",
    "- Protien-Protien-Tissue interaction\n",
    "- Tissue-Function-Gene interaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: networkx in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (3.4)\n",
      "Requirement already satisfied: pandas in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (2.2.3)\n",
      "Requirement already satisfied: arango in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (0.2.1)\n",
      "Collecting matplotlib\n",
      "  Obtaining dependency information for matplotlib from https://files.pythonhosted.org/packages/c7/8b/92e9da1f28310a1f6572b5c55097b0c0ceb5e27486d85fb73b54f5a9b939/matplotlib-3.10.0-cp311-cp311-macosx_11_0_arm64.whl.metadata\n",
      "  Downloading matplotlib-3.10.0-cp311-cp311-macosx_11_0_arm64.whl.metadata (11 kB)\n",
      "Requirement already satisfied: numpy>=1.23.2 in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from pandas) (1.26.4)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from pandas) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from pandas) (2025.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from pandas) (2025.1)\n",
      "Requirement already satisfied: nose in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from arango) (1.3.7)\n",
      "Requirement already satisfied: mock in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from arango) (5.1.0)\n",
      "Collecting contourpy>=1.0.1 (from matplotlib)\n",
      "  Obtaining dependency information for contourpy>=1.0.1 from https://files.pythonhosted.org/packages/67/71/1e6e95aee21a500415f5d2dbf037bf4567529b6a4e986594d7026ec5ae90/contourpy-1.3.1-cp311-cp311-macosx_11_0_arm64.whl.metadata\n",
      "  Downloading contourpy-1.3.1-cp311-cp311-macosx_11_0_arm64.whl.metadata (5.4 kB)\n",
      "Collecting cycler>=0.10 (from matplotlib)\n",
      "  Obtaining dependency information for cycler>=0.10 from https://files.pythonhosted.org/packages/e7/05/c19819d5e3d95294a6f5947fb9b9629efb316b96de511b418c53d245aae6/cycler-0.12.1-py3-none-any.whl.metadata\n",
      "  Downloading cycler-0.12.1-py3-none-any.whl.metadata (3.8 kB)\n",
      "Collecting fonttools>=4.22.0 (from matplotlib)\n",
      "  Obtaining dependency information for fonttools>=4.22.0 from https://files.pythonhosted.org/packages/35/56/a2f3e777d48fcae7ecd29de4d96352d84e5ea9871e5f3fc88241521572cf/fonttools-4.56.0-cp311-cp311-macosx_10_9_universal2.whl.metadata\n",
      "  Downloading fonttools-4.56.0-cp311-cp311-macosx_10_9_universal2.whl.metadata (101 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m101.9/101.9 kB\u001b[0m \u001b[31m2.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hCollecting kiwisolver>=1.3.1 (from matplotlib)\n",
      "  Obtaining dependency information for kiwisolver>=1.3.1 from https://files.pythonhosted.org/packages/5f/b4/c12b3ac0852a3a68f94598d4c8d569f55361beef6159dce4e7b624160da2/kiwisolver-1.4.8-cp311-cp311-macosx_11_0_arm64.whl.metadata\n",
      "  Downloading kiwisolver-1.4.8-cp311-cp311-macosx_11_0_arm64.whl.metadata (6.2 kB)\n",
      "Requirement already satisfied: packaging>=20.0 in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from matplotlib) (24.2)\n",
      "Collecting pillow>=8 (from matplotlib)\n",
      "  Obtaining dependency information for pillow>=8 from https://files.pythonhosted.org/packages/d9/45/3fe487010dd9ce0a06adf9b8ff4f273cc0a44536e234b0fad3532a42c15b/pillow-11.1.0-cp311-cp311-macosx_11_0_arm64.whl.metadata\n",
      "  Downloading pillow-11.1.0-cp311-cp311-macosx_11_0_arm64.whl.metadata (9.1 kB)\n",
      "Collecting pyparsing>=2.3.1 (from matplotlib)\n",
      "  Obtaining dependency information for pyparsing>=2.3.1 from https://files.pythonhosted.org/packages/1c/a7/c8a2d361bf89c0d9577c934ebb7421b25dc84bf3a8e3ac0a40aed9acc547/pyparsing-3.2.1-py3-none-any.whl.metadata\n",
      "  Downloading pyparsing-3.2.1-py3-none-any.whl.metadata (5.0 kB)\n",
      "Requirement already satisfied: six>=1.5 in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from python-dateutil>=2.8.2->pandas) (1.17.0)\n",
      "Downloading matplotlib-3.10.0-cp311-cp311-macosx_11_0_arm64.whl (8.0 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m8.0/8.0 MB\u001b[0m \u001b[31m1.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hDownloading contourpy-1.3.1-cp311-cp311-macosx_11_0_arm64.whl (254 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m254.5/254.5 kB\u001b[0m \u001b[31m1.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hDownloading cycler-0.12.1-py3-none-any.whl (8.3 kB)\n",
      "Downloading fonttools-4.56.0-cp311-cp311-macosx_10_9_universal2.whl (2.8 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.8/2.8 MB\u001b[0m \u001b[31m1.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hDownloading kiwisolver-1.4.8-cp311-cp311-macosx_11_0_arm64.whl (65 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m65.4/65.4 kB\u001b[0m \u001b[31m1.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hDownloading pillow-11.1.0-cp311-cp311-macosx_11_0_arm64.whl (3.1 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m1.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hDownloading pyparsing-3.2.1-py3-none-any.whl (107 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m107.7/107.7 kB\u001b[0m \u001b[31m1.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hInstalling collected packages: pyparsing, pillow, kiwisolver, fonttools, cycler, contourpy, matplotlib\n",
      "Successfully installed contourpy-1.3.1 cycler-0.12.1 fonttools-4.56.0 kiwisolver-1.4.8 matplotlib-3.10.0 pillow-11.1.0 pyparsing-3.2.1\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.2.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m25.0.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n",
      "Requirement already satisfied: nx-arangodb in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (1.3.0)\n",
      "Requirement already satisfied: networkx<=3.4,>=3.0 in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from nx-arangodb) (3.4)\n",
      "Requirement already satisfied: phenolrs~=0.5 in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from nx-arangodb) (0.5.9)\n",
      "Requirement already satisfied: python-arango~=8.1 in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from nx-arangodb) (8.1.4)\n",
      "Requirement already satisfied: adbnx-adapter~=5.0.5 in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from nx-arangodb) (5.0.6)\n",
      "Requirement already satisfied: requests>=2.27.1 in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from adbnx-adapter~=5.0.5->nx-arangodb) (2.32.3)\n",
      "Requirement already satisfied: rich>=12.5.1 in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from adbnx-adapter~=5.0.5->nx-arangodb) (13.9.4)\n",
      "Requirement already satisfied: setuptools>=45 in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from adbnx-adapter~=5.0.5->nx-arangodb) (65.5.0)\n",
      "Requirement already satisfied: numpy~=1.26 in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from phenolrs~=0.5->nx-arangodb) (1.26.4)\n",
      "Requirement already satisfied: urllib3>=1.26.0 in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from python-arango~=8.1->nx-arangodb) (2.3.0)\n",
      "Requirement already satisfied: requests_toolbelt in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from python-arango~=8.1->nx-arangodb) (1.0.0)\n",
      "Requirement already satisfied: PyJWT in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from python-arango~=8.1->nx-arangodb) (2.10.1)\n",
      "Requirement already satisfied: importlib_metadata>=4.7.1 in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from python-arango~=8.1->nx-arangodb) (8.6.1)\n",
      "Requirement already satisfied: packaging>=23.1 in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from python-arango~=8.1->nx-arangodb) (24.2)\n",
      "Requirement already satisfied: zipp>=3.20 in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from importlib_metadata>=4.7.1->python-arango~=8.1->nx-arangodb) (3.21.0)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from requests>=2.27.1->adbnx-adapter~=5.0.5->nx-arangodb) (3.4.1)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from requests>=2.27.1->adbnx-adapter~=5.0.5->nx-arangodb) (3.10)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from requests>=2.27.1->adbnx-adapter~=5.0.5->nx-arangodb) (2025.1.31)\n",
      "Requirement already satisfied: markdown-it-py>=2.2.0 in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from rich>=12.5.1->adbnx-adapter~=5.0.5->nx-arangodb) (3.0.0)\n",
      "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from rich>=12.5.1->adbnx-adapter~=5.0.5->nx-arangodb) (2.19.1)\n",
      "Requirement already satisfied: mdurl~=0.1 in /Users/redomic/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages (from markdown-it-py>=2.2.0->rich>=12.5.1->adbnx-adapter~=5.0.5->nx-arangodb) (0.1.2)\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.2.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m25.0.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip install networkx pandas arango matplotlib\n",
    "!pip install nx-arangodb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2025-02-24 13:31:42--  https://snap.stanford.edu/biodata/datasets/10032/files/GG-NE.tar.gz\n",
      "Resolving snap.stanford.edu (snap.stanford.edu)... 171.64.75.80\n",
      "Connecting to snap.stanford.edu (snap.stanford.edu)|171.64.75.80|:443... connected.\n",
      "HTTP request sent, awaiting response... 206 Partial Content\n",
      "Length: 20134095145 (19G), 1593698601 (1.5G) remaining [application/x-gzip]\n",
      "Saving to: ‘GG-NE.tar.gz’\n",
      "\n",
      "GG-NE.tar.gz        100%[++++++++++++++++++=>]  18.75G   748KB/s    in 35m 6s  \n",
      "\n",
      "2025-02-24 14:06:49 (739 KB/s) - ‘GG-NE.tar.gz’ saved [20134095145/20134095145]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Interactions\n",
    "\n",
    "!wget -nc https://snap.stanford.edu/biodata/datasets/10001/files/ChCh-Miner_durgbank-chem-chem.tsv.gz\n",
    "!wget -nc https://snap.stanford.edu/biodata/datasets/10002/files/ChG-Miner_miner-chem-gene.tsv.gz\n",
    "!wget -nc https://snap.stanford.edu/biodata/datasets/10004/files/DCh-Miner_miner-disease-chemical.tsv.gz\n",
    "!wget -nc https://snap.stanford.edu/biodata/datasets/10006/files/DD-Miner_miner-disease-disease.tsv.gz\n",
    "!wget -nc https://snap.stanford.edu/biodata/datasets/10019/files/DF-Miner_miner-disease-function.tsv.gz\n",
    "!wget -nc https://snap.stanford.edu/biodata/datasets/10020/files/DG-Miner_miner-disease-gene.tsv.gz\n",
    "!wget -nc https://snap.stanford.edu/biodata/datasets/10026/files/FF-Miner_miner-func-func.tsv.gz\n",
    "!wget -nc https://snap.stanford.edu/biodata/datasets/10024/files/GF-Miner_miner-gene-function.tsv.gz\n",
    "!wget -nc https://snap.stanford.edu/biodata/datasets/10027/files/GP-Miner_miner-gene-protein.tsv.gz\n",
    "#!wget -nc https://snap.stanford.edu/biodata/datasets/10028/files/PP-Miner_miner-ppi.tsv.gz\n",
    "#!wget -nc https://snap.stanford.edu/biodata/datasets/10032/files/GG-NE.tar.gz\n",
    "\n",
    "# Entities and Feature tables\n",
    "\n",
    "!wget -nc https://snap.stanford.edu/biodata/datasets/10021/files/D-DoMiner_miner-diseaseDOID.tsv.gz\n",
    "!wget -nc https://snap.stanford.edu/biodata/datasets/10003/files/D-MeshMiner_miner-disease.tsv.gz\n",
    "!wget -nc https://snap.stanford.edu/biodata/datasets/10025/files/D-OmimMiner_miner-diseaseOMIM.tsv.gz\n",
    "!wget -nc https://snap.stanford.edu/biodata/datasets/10022/files/G-SynMiner_miner-geneHUGO.tsv.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[19:08:33 +0530] [INFO]: NetworkX-cuGraph is unavailable: No module named 'cupy'.\n"
     ]
    }
   ],
   "source": [
    "import networkx as nx\n",
    "import pandas as pd\n",
    "import nx_arangodb as nxadb\n",
    "import requests\n",
    "from arango import ArangoClient\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "\n",
    "os.environ[\"DATABASE_HOST\"] = \"http://localhost:8529\"  # Replace with your ArangoDB host\n",
    "os.environ[\"DATABASE_USERNAME\"] = \"root\"               # Replace with your ArangoDB username\n",
    "os.environ[\"DATABASE_PASSWORD\"] = \"openSesame\"         # Replace with your ArangoDB password\n",
    "os.environ[\"DATABASE_NAME\"] = \"NeuThera\"               "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading Dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Interactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug</th>\n",
       "      <th>drug_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DB00862</td>\n",
       "      <td>DB00966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DB00575</td>\n",
       "      <td>DB00806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DB01242</td>\n",
       "      <td>DB08893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DB01151</td>\n",
       "      <td>DB08883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DB01235</td>\n",
       "      <td>DB01275</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      drug drug_target\n",
       "0  DB00862     DB00966\n",
       "1  DB00575     DB00806\n",
       "2  DB01242     DB08893\n",
       "3  DB01151     DB08883\n",
       "4  DB01235     DB01275"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drug-Drug interaction\n",
    "\n",
    "ChCh = pd.read_csv(\n",
    "    \"./ChCh-Miner_durgbank-chem-chem.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    names=[\"drug\", \"drug_target\"],\n",
    ")\n",
    "\n",
    "ChCh.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DB00357</td>\n",
       "      <td>P05108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DB02721</td>\n",
       "      <td>P00325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DB00773</td>\n",
       "      <td>P23219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DB07138</td>\n",
       "      <td>Q16539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DB08136</td>\n",
       "      <td>P24941</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      drug    gene\n",
       "0  DB00357  P05108\n",
       "1  DB02721  P00325\n",
       "2  DB00773  P23219\n",
       "3  DB07138  Q16539\n",
       "4  DB08136  P24941"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drug-Gene interaction\n",
    "\n",
    "ChG = pd.read_csv(\n",
    "    \"./ChG-Miner_miner-chem-gene.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    header=0,\n",
    "    names=[\"drug\", \"gene\"],\n",
    ")\n",
    "\n",
    "ChG.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mesh</th>\n",
       "      <th>drug</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MESH:D005923</td>\n",
       "      <td>DB00564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MESH:D009503</td>\n",
       "      <td>DB01072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MESH:D016115</td>\n",
       "      <td>DB01759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MESH:D018476</td>\n",
       "      <td>DB00451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MESH:C567059</td>\n",
       "      <td>DB00641</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           mesh     drug\n",
       "0  MESH:D005923  DB00564\n",
       "1  MESH:D009503  DB01072\n",
       "2  MESH:D016115  DB01759\n",
       "3  MESH:D018476  DB00451\n",
       "4  MESH:C567059  DB00641"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Disease-Drug interaction\n",
    "\n",
    "DCh = pd.read_csv(\n",
    "    \"./DCh-Miner_miner-disease-chemical.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    header=0,\n",
    "    names=[\"mesh\", \"drug\"],\n",
    ")\n",
    "\n",
    "DCh.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>doid</th>\n",
       "      <th>doid_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DOID:0001816</td>\n",
       "      <td>DOID:1115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DOID:0002116</td>\n",
       "      <td>DOID:10124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DOID:0014667</td>\n",
       "      <td>DOID:4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DOID:0050004</td>\n",
       "      <td>DOID:10400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DOID:0050012</td>\n",
       "      <td>DOID:934</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           doid doid_target\n",
       "0  DOID:0001816   DOID:1115\n",
       "1  DOID:0002116  DOID:10124\n",
       "2  DOID:0014667      DOID:4\n",
       "3  DOID:0050004  DOID:10400\n",
       "4  DOID:0050012    DOID:934"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Disease-Disease interaction\n",
    "\n",
    "DD = pd.read_csv(\n",
    "    \"./DD-Miner_miner-disease-disease.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    header=0,\n",
    "    names=[\"doid\", \"doid_target\"],\n",
    ")\n",
    "\n",
    "DD.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mesh</th>\n",
       "      <th>go</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MESH:D000037</td>\n",
       "      <td>GO:0009257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MESH:C536409</td>\n",
       "      <td>GO:0009257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MESH:D009436</td>\n",
       "      <td>GO:0009257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MESH:D000860</td>\n",
       "      <td>GO:0009258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MESH:D008106</td>\n",
       "      <td>GO:0009258</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           mesh          go\n",
       "0  MESH:D000037  GO:0009257\n",
       "1  MESH:C536409  GO:0009257\n",
       "2  MESH:D009436  GO:0009257\n",
       "3  MESH:D000860  GO:0009258\n",
       "4  MESH:D008106  GO:0009258"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Disease-Function interaction\n",
    "\n",
    "DF = pd.read_csv(\n",
    "    \"./DF-Miner_miner-disease-function.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    header=0,\n",
    "    names=[\"mesh\", \"go\"],\n",
    ")\n",
    "\n",
    "DF.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mesh</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MESH:D005756</td>\n",
       "      <td>A0A087WZV0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MESH:D055370</td>\n",
       "      <td>P11464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MESH:D007410</td>\n",
       "      <td>Q92945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MESH:D014062</td>\n",
       "      <td>Q6ISS4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MESH:D054549</td>\n",
       "      <td>Q96RU8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           mesh        gene\n",
       "0  MESH:D005756  A0A087WZV0\n",
       "1  MESH:D055370      P11464\n",
       "2  MESH:D007410      Q92945\n",
       "3  MESH:D014062      Q6ISS4\n",
       "4  MESH:D054549      Q96RU8"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Disease-Gene interaction\n",
    "\n",
    "DG = pd.read_csv(\n",
    "    \"./DG-Miner_miner-disease-gene.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    header=0,\n",
    "    names=[\"mesh\", \"gene\"],\n",
    ")\n",
    "\n",
    "DG.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>go</th>\n",
       "      <th>go_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0008296</td>\n",
       "      <td>GO:0008408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0016811</td>\n",
       "      <td>GO:0033970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0045222</td>\n",
       "      <td>GO:0045223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0021803</td>\n",
       "      <td>GO:0030031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0033574</td>\n",
       "      <td>GO:1901654</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           go   go_target\n",
       "0  GO:0008296  GO:0008408\n",
       "1  GO:0016811  GO:0033970\n",
       "2  GO:0045222  GO:0045223\n",
       "3  GO:0021803  GO:0030031\n",
       "4  GO:0033574  GO:1901654"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Function-Function interaction\n",
    "\n",
    "FF = pd.read_csv(\n",
    "    \"./FF-Miner_miner-func-func.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    header=0,\n",
    "    names=[\"go\", \"go_target\"],\n",
    ")\n",
    "\n",
    "FF.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>go</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0005509</td>\n",
       "      <td>PDCD6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0004672</td>\n",
       "      <td>CDK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0005524</td>\n",
       "      <td>CDK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0005634</td>\n",
       "      <td>CDK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0005737</td>\n",
       "      <td>CDK1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           go   gene\n",
       "0  GO:0005509  PDCD6\n",
       "1  GO:0004672   CDK1\n",
       "2  GO:0005524   CDK1\n",
       "3  GO:0005634   CDK1\n",
       "4  GO:0005737   CDK1"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Gene-Function interaction (ONLY HUMANS)\n",
    "\n",
    "\n",
    "GF = pd.read_csv(\n",
    "    \"./GF-Miner_miner-gene-function.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    skipinitialspace=True\n",
    ")\n",
    "\n",
    "GF = GF.rename(\n",
    "    columns={\n",
    "        \"# GO_ID\": \"go\",\n",
    "        \"Gene\": \"gene\",\n",
    "        \"C8\": \"go_category\",\n",
    "        \"C10\": \"protein\",\n",
    "        \"C12\": \"organism\",\n",
    "        \"C13\": \"date\"\n",
    "    }\n",
    ")\n",
    "\n",
    "GF = GF[GF[\"organism\"] == \"taxon:9606\"]\n",
    "GF = GF[[\"go\", \"gene\"]]\n",
    "\n",
    "GF.head(5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensg</th>\n",
       "      <th>ensp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>ENSP00000354687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENSP00000355046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000198804</td>\n",
       "      <td>ENSP00000354499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000198712</td>\n",
       "      <td>ENSP00000354876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000228253</td>\n",
       "      <td>ENSP00000355265</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              ensg             ensp\n",
       "0  ENSG00000198888  ENSP00000354687\n",
       "1  ENSG00000198763  ENSP00000355046\n",
       "2  ENSG00000198804  ENSP00000354499\n",
       "3  ENSG00000198712  ENSP00000354876\n",
       "4  ENSG00000228253  ENSP00000355265"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Gene-Protien interaction\n",
    "\n",
    "GP = pd.read_csv(\n",
    "    \"./GP-Miner_miner-gene-protein.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    header=0,\n",
    "    names=[\"ensg\", \"ensp\"],\n",
    ")\n",
    "\n",
    "GP.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Protien-Protien interaction\n",
    "\n",
    "# PP = pd.read_csv(\n",
    "#     \"./GG-NE.tar.gz\",\n",
    "#     compression=\"gzip\",\n",
    "#     sep=\"\\t\",\n",
    "#     nrows=5,\n",
    "#     # header=0,\n",
    "#     # names=[\"ensg\", \"ensp\"],\n",
    "# )\n",
    "\n",
    "# PP.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Entities and Feature Tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>doid</th>\n",
       "      <th>name</th>\n",
       "      <th>definition</th>\n",
       "      <th>synonym</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DOID:0001816</td>\n",
       "      <td>angiosarcoma</td>\n",
       "      <td>A malignant vascular tumor that results_in rap...</td>\n",
       "      <td>hemangiosarcoma EXACT []</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DOID:0002116</td>\n",
       "      <td>pterygium</td>\n",
       "      <td>NaN</td>\n",
       "      <td>surfer's eye EXACT []</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DOID:0014667</td>\n",
       "      <td>disease of metabolism</td>\n",
       "      <td>A disease that involving errors in metabolic p...</td>\n",
       "      <td>metabolic disease  EXACT [SNOMEDCT_2005_07_31:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DOID:0050001</td>\n",
       "      <td>Actinomadura madurae infectious disease</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DOID:0050002</td>\n",
       "      <td>Actinomadura pelletieri infectious disease</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9242</th>\n",
       "      <td>DOID:9989</td>\n",
       "      <td>metastasis to the orbit</td>\n",
       "      <td>NaN</td>\n",
       "      <td>secondary malignant neoplasm of orbit (disorde...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9243</th>\n",
       "      <td>DOID:999</td>\n",
       "      <td>eosinophilia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eosinophilic leukocytosis EXACT [MTHICD9_2006:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9244</th>\n",
       "      <td>DOID:9993</td>\n",
       "      <td>hypoglycemia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hypoglycaemia EXACT [SNOMEDCT_2005_07_31:15469...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9245</th>\n",
       "      <td>DOID:9995</td>\n",
       "      <td>endocrine and metabolic disturbances specific ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9246</th>\n",
       "      <td>DOID:9997</td>\n",
       "      <td>peripartum cardiomyopathy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>postpartum peripartum cardiomyopathy EXACT []</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9247 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              doid                                               name  \\\n",
       "0     DOID:0001816                                       angiosarcoma   \n",
       "1     DOID:0002116                                          pterygium   \n",
       "2     DOID:0014667                              disease of metabolism   \n",
       "3     DOID:0050001            Actinomadura madurae infectious disease   \n",
       "4     DOID:0050002         Actinomadura pelletieri infectious disease   \n",
       "...            ...                                                ...   \n",
       "9242     DOID:9989                            metastasis to the orbit   \n",
       "9243      DOID:999                                       eosinophilia   \n",
       "9244     DOID:9993                                       hypoglycemia   \n",
       "9245     DOID:9995  endocrine and metabolic disturbances specific ...   \n",
       "9246     DOID:9997                          peripartum cardiomyopathy   \n",
       "\n",
       "                                             definition  \\\n",
       "0     A malignant vascular tumor that results_in rap...   \n",
       "1                                                   NaN   \n",
       "2     A disease that involving errors in metabolic p...   \n",
       "3                                                   NaN   \n",
       "4                                                   NaN   \n",
       "...                                                 ...   \n",
       "9242                                                NaN   \n",
       "9243                                                NaN   \n",
       "9244                                                NaN   \n",
       "9245                                                NaN   \n",
       "9246                                                NaN   \n",
       "\n",
       "                                                synonym  \n",
       "0                              hemangiosarcoma EXACT []  \n",
       "1                                 surfer's eye EXACT []  \n",
       "2     metabolic disease  EXACT [SNOMEDCT_2005_07_31:...  \n",
       "3                                                   NaN  \n",
       "4                                                   NaN  \n",
       "...                                                 ...  \n",
       "9242  secondary malignant neoplasm of orbit (disorde...  \n",
       "9243  Eosinophilic leukocytosis EXACT [MTHICD9_2006:...  \n",
       "9244  Hypoglycaemia EXACT [SNOMEDCT_2005_07_31:15469...  \n",
       "9245                                                NaN  \n",
       "9246      postpartum peripartum cardiomyopathy EXACT []  \n",
       "\n",
       "[9247 rows x 4 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Disease DOID Synopses\n",
    "\n",
    "Doid = pd.read_csv(\n",
    "    \"./D-DoMiner_miner-diseaseDOID.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    header=0,\n",
    "    names=[\"doid\", \"name\", \"definition\", \"synonym\"],\n",
    ")\n",
    "\n",
    "Doid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mesh</th>\n",
       "      <th>name</th>\n",
       "      <th>definition</th>\n",
       "      <th>synonym</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MESH:C538288</td>\n",
       "      <td>10p Deletion Syndrome (Partial)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Chromosome 10, 10p- Partial|Chromosome 10, mon...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MESH:C535484</td>\n",
       "      <td>13q deletion syndrome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Chromosome 13q deletion|Chromosome 13q deletio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MESH:C579849</td>\n",
       "      <td>15q24 Microdeletion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15q24 Deletion|15q24 Microdeletion Syndrome|In...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MESH:C579850</td>\n",
       "      <td>16p11.2 Deletion Syndrome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MESH:C567076</td>\n",
       "      <td>17,20-Lyase Deficiency, Isolated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17-Alpha-Hydroxylase-17,20-Lyase Deficiency, C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11327</th>\n",
       "      <td>MESH:C536729</td>\n",
       "      <td>Zunich neuroectodermal syndrome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHIME syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11328</th>\n",
       "      <td>MESH:C536730</td>\n",
       "      <td>Zuska's Disease</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lactation and squamous metaplasia of lactifero...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11329</th>\n",
       "      <td>MESH:C565223</td>\n",
       "      <td>Zygodactyly 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11330</th>\n",
       "      <td>MESH:D015051</td>\n",
       "      <td>Zygomatic Fractures</td>\n",
       "      <td>Fractures of the zygoma.</td>\n",
       "      <td>Fractures, Zygomatic|Fracture, Zygomatic|Zygom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11331</th>\n",
       "      <td>MESH:D020096</td>\n",
       "      <td>Zygomycosis</td>\n",
       "      <td>Infection in humans and animals caused by fung...</td>\n",
       "      <td>Entomophthoramycoses|Entomophthoramycosis|Phyc...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11332 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               mesh                              name  \\\n",
       "0      MESH:C538288   10p Deletion Syndrome (Partial)   \n",
       "1      MESH:C535484             13q deletion syndrome   \n",
       "2      MESH:C579849               15q24 Microdeletion   \n",
       "3      MESH:C579850         16p11.2 Deletion Syndrome   \n",
       "4      MESH:C567076  17,20-Lyase Deficiency, Isolated   \n",
       "...             ...                               ...   \n",
       "11327  MESH:C536729   Zunich neuroectodermal syndrome   \n",
       "11328  MESH:C536730                   Zuska's Disease   \n",
       "11329  MESH:C565223                     Zygodactyly 1   \n",
       "11330  MESH:D015051               Zygomatic Fractures   \n",
       "11331  MESH:D020096                       Zygomycosis   \n",
       "\n",
       "                                              definition  \\\n",
       "0                                                    NaN   \n",
       "1                                                    NaN   \n",
       "2                                                    NaN   \n",
       "3                                                    NaN   \n",
       "4                                                    NaN   \n",
       "...                                                  ...   \n",
       "11327                                                NaN   \n",
       "11328                                                NaN   \n",
       "11329                                                NaN   \n",
       "11330                           Fractures of the zygoma.   \n",
       "11331  Infection in humans and animals caused by fung...   \n",
       "\n",
       "                                                 synonym  \n",
       "0      Chromosome 10, 10p- Partial|Chromosome 10, mon...  \n",
       "1      Chromosome 13q deletion|Chromosome 13q deletio...  \n",
       "2      15q24 Deletion|15q24 Microdeletion Syndrome|In...  \n",
       "3                                                    NaN  \n",
       "4      17-Alpha-Hydroxylase-17,20-Lyase Deficiency, C...  \n",
       "...                                                  ...  \n",
       "11327                                     CHIME syndrome  \n",
       "11328  Lactation and squamous metaplasia of lactifero...  \n",
       "11329                                                ZD1  \n",
       "11330  Fractures, Zygomatic|Fracture, Zygomatic|Zygom...  \n",
       "11331  Entomophthoramycoses|Entomophthoramycosis|Phyc...  \n",
       "\n",
       "[11332 rows x 4 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Disease MESH Synopses\n",
    "\n",
    "Mesh = pd.read_csv(\n",
    "    \"./D-MeshMiner_miner-disease.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    header=0,\n",
    "    names=[\"mesh\", \"name\", \"definition\", \"synonym\"],\n",
    ")\n",
    "\n",
    "Mesh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>omim</th>\n",
       "      <th>phenotypes</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene</th>\n",
       "      <th>location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OMIM:115665</td>\n",
       "      <td>Cataract 8, multiple types (2)</td>\n",
       "      <td>Cataract, congenital, Volkmann type</td>\n",
       "      <td>CTRCT8, CCV</td>\n",
       "      <td>1pter-p36.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OMIM:607671</td>\n",
       "      <td>Dystonia 13, torsion (2)</td>\n",
       "      <td>Dystonia 13, torsion</td>\n",
       "      <td>DYT13</td>\n",
       "      <td>1p36.32-p36.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OMIM:606242</td>\n",
       "      <td>Kondoh syndrome (2)</td>\n",
       "      <td>Kondoh syndrome (mental retardation, microceph...</td>\n",
       "      <td>KONDS</td>\n",
       "      <td>1p36.32-p35.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OMIM:614414</td>\n",
       "      <td>Deafness, autosomal recessive 96 (2)</td>\n",
       "      <td>Deafness, autosomal recessive 96</td>\n",
       "      <td>DFNB96</td>\n",
       "      <td>1p36.31-p36.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OMIM:609918</td>\n",
       "      <td>Gallbladder disease 2 (2)</td>\n",
       "      <td>Gallbladder disease 2</td>\n",
       "      <td>GBD2</td>\n",
       "      <td>1p36.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1186</th>\n",
       "      <td>OMIM:300519</td>\n",
       "      <td>Mental retardation, X-linked, syndromic, Marti...</td>\n",
       "      <td>Mental retardation, X-linked, syndromic, Marti...</td>\n",
       "      <td>MRXSMP</td>\n",
       "      <td>Chr.X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1187</th>\n",
       "      <td>OMIM:400042</td>\n",
       "      <td>Spermatogenic failure, Y-linked, 1 (4)</td>\n",
       "      <td>Chromosome Yq11 interstitial deletion syndrome</td>\n",
       "      <td>DELYq11, CYDELq11, SPGFY1</td>\n",
       "      <td>Yq11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1188</th>\n",
       "      <td>OMIM:475000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Growth control, Y-chromosome influenced</td>\n",
       "      <td>GCY, TSY, STA</td>\n",
       "      <td>Yq12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1189</th>\n",
       "      <td>OMIM:400043</td>\n",
       "      <td>Deafness, Y-linked 1 (1)</td>\n",
       "      <td>Deafness, Y-linked 1</td>\n",
       "      <td>DFNY1</td>\n",
       "      <td>Chr.Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1190</th>\n",
       "      <td>OMIM:400004</td>\n",
       "      <td>Retinitis pigmentosa, Y-linked (2)</td>\n",
       "      <td>Retinitis pigmentosa, Y-linked</td>\n",
       "      <td>RPY</td>\n",
       "      <td>Chr.Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1191 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             omim                                         phenotypes  \\\n",
       "0     OMIM:115665                     Cataract 8, multiple types (2)   \n",
       "1     OMIM:607671                           Dystonia 13, torsion (2)   \n",
       "2     OMIM:606242                                Kondoh syndrome (2)   \n",
       "3     OMIM:614414               Deafness, autosomal recessive 96 (2)   \n",
       "4     OMIM:609918                          Gallbladder disease 2 (2)   \n",
       "...           ...                                                ...   \n",
       "1186  OMIM:300519  Mental retardation, X-linked, syndromic, Marti...   \n",
       "1187  OMIM:400042             Spermatogenic failure, Y-linked, 1 (4)   \n",
       "1188  OMIM:475000                                                NaN   \n",
       "1189  OMIM:400043                           Deafness, Y-linked 1 (1)   \n",
       "1190  OMIM:400004                 Retinitis pigmentosa, Y-linked (2)   \n",
       "\n",
       "                                              gene_name  \\\n",
       "0                   Cataract, congenital, Volkmann type   \n",
       "1                                  Dystonia 13, torsion   \n",
       "2     Kondoh syndrome (mental retardation, microceph...   \n",
       "3                      Deafness, autosomal recessive 96   \n",
       "4                                 Gallbladder disease 2   \n",
       "...                                                 ...   \n",
       "1186  Mental retardation, X-linked, syndromic, Marti...   \n",
       "1187     Chromosome Yq11 interstitial deletion syndrome   \n",
       "1188            Growth control, Y-chromosome influenced   \n",
       "1189                               Deafness, Y-linked 1   \n",
       "1190                     Retinitis pigmentosa, Y-linked   \n",
       "\n",
       "                           gene        location  \n",
       "0                   CTRCT8, CCV    1pter-p36.13  \n",
       "1                         DYT13  1p36.32-p36.13  \n",
       "2                         KONDS   1p36.32-p35.3  \n",
       "3                        DFNB96  1p36.31-p36.13  \n",
       "4                          GBD2         1p36.21  \n",
       "...                         ...             ...  \n",
       "1186                     MRXSMP           Chr.X  \n",
       "1187  DELYq11, CYDELq11, SPGFY1            Yq11  \n",
       "1188              GCY, TSY, STA            Yq12  \n",
       "1189                      DFNY1           Chr.Y  \n",
       "1190                        RPY           Chr.Y  \n",
       "\n",
       "[1191 rows x 5 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# OMIM Genetic Disorders\n",
    "\n",
    "Omim = pd.read_csv(\n",
    "    \"./D-OmimMiner_miner-diseaseOMIM.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    header=0,\n",
    "    names=[\"omim\", \"phenotypes\", \"gene_name\", \"gene\", \"location\", \"_\"],\n",
    ")\n",
    "\n",
    "Omim = Omim.iloc[:, :-1]\n",
    "\n",
    "Omim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/xb/gzq2kh212g9g01hg900w3y1r0000gn/T/ipykernel_87442/3605880204.py:3: DtypeWarning: Columns (32,34,38) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  Gene = pd.read_csv(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensg</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>locus_group</th>\n",
       "      <th>locus_type</th>\n",
       "      <th>status</th>\n",
       "      <th>location</th>\n",
       "      <th>location_sortable</th>\n",
       "      <th>alias_symbol</th>\n",
       "      <th>...</th>\n",
       "      <th>horde_id</th>\n",
       "      <th>merops</th>\n",
       "      <th>imgt</th>\n",
       "      <th>iuphar</th>\n",
       "      <th>kznf_gene_catalog</th>\n",
       "      <th>mamit-trnadb</th>\n",
       "      <th>cd</th>\n",
       "      <th>lncrnadb</th>\n",
       "      <th>enzyme_id</th>\n",
       "      <th>intermediate_filament_db</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>I43.950</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000268895</td>\n",
       "      <td>HGNC:37133</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>A1BG antisense RNA 1</td>\n",
       "      <td>non-coding RNA</td>\n",
       "      <td>RNA, long non-coding</td>\n",
       "      <td>Approved</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>FLJ23569</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>HGNC:24086</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>10q21.1</td>\n",
       "      <td>10q21.1</td>\n",
       "      <td>ACF|ASP|ACF64|ACF65|APOBEC1CF</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>FWP007|S863-7|CPAMD5</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>I39.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000245105</td>\n",
       "      <td>HGNC:27057</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>A2M antisense RNA 1 (head to head)</td>\n",
       "      <td>non-coding RNA</td>\n",
       "      <td>RNA, long non-coding</td>\n",
       "      <td>Approved</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35649</th>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>HGNC:25820</td>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>zyg-11 family member B, cell cycle regulator</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1p32.3</td>\n",
       "      <td>01p32.3</td>\n",
       "      <td>FLJ13456</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35650</th>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>HGNC:13200</td>\n",
       "      <td>ZYX</td>\n",
       "      <td>zyxin</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>7q32</td>\n",
       "      <td>07q32</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35651</th>\n",
       "      <td>ENSG00000274572</td>\n",
       "      <td>HGNC:51695</td>\n",
       "      <td>ZYXP1</td>\n",
       "      <td>zyxin pseudogene 1</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>Approved</td>\n",
       "      <td>8q24.23</td>\n",
       "      <td>08q24.23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35652</th>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>HGNC:29027</td>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>zinc finger ZZ-type and EF-hand domain contain...</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>17p13.3</td>\n",
       "      <td>17p13.3</td>\n",
       "      <td>KIAA0399|ZZZ4|FLJ10821</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35653</th>\n",
       "      <td>ENSG00000036549</td>\n",
       "      <td>HGNC:24523</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>zinc finger ZZ-type containing 3</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1p31.1</td>\n",
       "      <td>01p31.1</td>\n",
       "      <td>DKFZP564I052|ATAC1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35654 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  ensg     hgnc_id      gene  \\\n",
       "0      ENSG00000121410      HGNC:5      A1BG   \n",
       "1      ENSG00000268895  HGNC:37133  A1BG-AS1   \n",
       "2      ENSG00000148584  HGNC:24086      A1CF   \n",
       "3      ENSG00000175899      HGNC:7       A2M   \n",
       "4      ENSG00000245105  HGNC:27057   A2M-AS1   \n",
       "...                ...         ...       ...   \n",
       "35649  ENSG00000162378  HGNC:25820    ZYG11B   \n",
       "35650  ENSG00000159840  HGNC:13200       ZYX   \n",
       "35651  ENSG00000274572  HGNC:51695     ZYXP1   \n",
       "35652  ENSG00000074755  HGNC:29027     ZZEF1   \n",
       "35653  ENSG00000036549  HGNC:24523      ZZZ3   \n",
       "\n",
       "                                               gene_name          locus_group  \\\n",
       "0                                 alpha-1-B glycoprotein  protein-coding gene   \n",
       "1                                   A1BG antisense RNA 1       non-coding RNA   \n",
       "2                         APOBEC1 complementation factor  protein-coding gene   \n",
       "3                                  alpha-2-macroglobulin  protein-coding gene   \n",
       "4                     A2M antisense RNA 1 (head to head)       non-coding RNA   \n",
       "...                                                  ...                  ...   \n",
       "35649       zyg-11 family member B, cell cycle regulator  protein-coding gene   \n",
       "35650                                              zyxin  protein-coding gene   \n",
       "35651                                 zyxin pseudogene 1           pseudogene   \n",
       "35652  zinc finger ZZ-type and EF-hand domain contain...  protein-coding gene   \n",
       "35653                   zinc finger ZZ-type containing 3  protein-coding gene   \n",
       "\n",
       "                      locus_type    status  location location_sortable  \\\n",
       "0      gene with protein product  Approved  19q13.43          19q13.43   \n",
       "1           RNA, long non-coding  Approved  19q13.43          19q13.43   \n",
       "2      gene with protein product  Approved   10q21.1           10q21.1   \n",
       "3      gene with protein product  Approved  12p13.31          12p13.31   \n",
       "4           RNA, long non-coding  Approved  12p13.31          12p13.31   \n",
       "...                          ...       ...       ...               ...   \n",
       "35649  gene with protein product  Approved    1p32.3           01p32.3   \n",
       "35650  gene with protein product  Approved      7q32             07q32   \n",
       "35651                 pseudogene  Approved   8q24.23          08q24.23   \n",
       "35652  gene with protein product  Approved   17p13.3           17p13.3   \n",
       "35653  gene with protein product  Approved    1p31.1           01p31.1   \n",
       "\n",
       "                        alias_symbol  ... horde_id   merops imgt iuphar  \\\n",
       "0                                NaN  ...      NaN  I43.950  NaN    NaN   \n",
       "1                           FLJ23569  ...      NaN      NaN  NaN    NaN   \n",
       "2      ACF|ASP|ACF64|ACF65|APOBEC1CF  ...      NaN      NaN  NaN    NaN   \n",
       "3               FWP007|S863-7|CPAMD5  ...      NaN  I39.001  NaN    NaN   \n",
       "4                                NaN  ...      NaN      NaN  NaN    NaN   \n",
       "...                              ...  ...      ...      ...  ...    ...   \n",
       "35649                       FLJ13456  ...      NaN      NaN  NaN    NaN   \n",
       "35650                            NaN  ...      NaN      NaN  NaN    NaN   \n",
       "35651                            NaN  ...      NaN      NaN  NaN    NaN   \n",
       "35652         KIAA0399|ZZZ4|FLJ10821  ...      NaN      NaN  NaN    NaN   \n",
       "35653             DKFZP564I052|ATAC1  ...      NaN      NaN  NaN    NaN   \n",
       "\n",
       "      kznf_gene_catalog mamit-trnadb   cd lncrnadb enzyme_id  \\\n",
       "0                   NaN          NaN  NaN      NaN       NaN   \n",
       "1                   NaN          NaN  NaN      NaN       NaN   \n",
       "2                   NaN          NaN  NaN      NaN       NaN   \n",
       "3                   NaN          NaN  NaN      NaN       NaN   \n",
       "4                   NaN          NaN  NaN      NaN       NaN   \n",
       "...                 ...          ...  ...      ...       ...   \n",
       "35649               NaN          NaN  NaN      NaN       NaN   \n",
       "35650               NaN          NaN  NaN      NaN       NaN   \n",
       "35651               NaN          NaN  NaN      NaN       NaN   \n",
       "35652               NaN          NaN  NaN      NaN       NaN   \n",
       "35653               NaN          NaN  NaN      NaN       NaN   \n",
       "\n",
       "       intermediate_filament_db  \n",
       "0                           NaN  \n",
       "1                           NaN  \n",
       "2                           NaN  \n",
       "3                           NaN  \n",
       "4                           NaN  \n",
       "...                         ...  \n",
       "35649                       NaN  \n",
       "35650                       NaN  \n",
       "35651                       NaN  \n",
       "35652                       NaN  \n",
       "35653                       NaN  \n",
       "\n",
       "[35654 rows x 48 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Gene\n",
    "\n",
    "Gene = pd.read_csv(\n",
    "    \"./G-SynMiner_miner-geneHUGO.tsv.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "\n",
    "Gene = Gene.rename(columns={'# ensembl_gene_id': 'ensg', 'symbol': 'gene', 'name': 'gene_name'})\n",
    "\n",
    "Gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/xb/gzq2kh212g9g01hg900w3y1r0000gn/T/ipykernel_23041/4251020720.py:11: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.\n",
      "  ChemRepresentation = pd.read_csv(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chembl_id\\tcanonical_smiles\\tstandard_inchi\\tstandard_inchi_key</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHEMBL153534\\tCc1cc(-c2csc(N=C(N)N)n2)cn1C\\tIn...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHEMBL440060\\tCC[C@H](C)[C@H](NC(=O)[C@H](CC(C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHEMBL440245\\tCCCC[C@@H]1NC(=O)[C@@H](NC(=O)[C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHEMBL440249\\tCC(C)C[C@@H]1NC(=O)CNC(=O)[C@H](...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHEMBL405398\\tBrc1cccc(Nc2ncnc3ccncc23)c1NCCN1...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chembl_id\\tcanonical_smiles\\tstandard_inchi\\tstandard_inchi_key\n",
       "0  CHEMBL153534\\tCc1cc(-c2csc(N=C(N)N)n2)cn1C\\tIn...             \n",
       "1  CHEMBL440060\\tCC[C@H](C)[C@H](NC(=O)[C@H](CC(C...             \n",
       "2  CHEMBL440245\\tCCCC[C@@H]1NC(=O)[C@@H](NC(=O)[C...             \n",
       "3  CHEMBL440249\\tCC(C)C[C@@H]1NC(=O)CNC(=O)[C@H](...             \n",
       "4  CHEMBL405398\\tBrc1cccc(Nc2ncnc3ccncc23)c1NCCN1...             "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# DrugBank\n",
    "\n",
    "Drug = pd.read_csv(\n",
    "    \"./drugbank_all_drugbank_vocabulary.csv.zip\",\n",
    "    compression=\"zip\",\n",
    "    sep=\",\",\n",
    "    header=0,\n",
    "    names=[\"drug\", \"accession\", \"drug_name\", \"cas\", \"unii\", \"synonym\", \"key\"],\n",
    ")\n",
    "\n",
    "ChemRepresentation = pd.read_csv(\n",
    "    \"./chembl_35_chemreps.txt.gz\",\n",
    "    compression=\"gzip\",\n",
    "    sep=\"\\t\",\n",
    "    # header=0,\n",
    "    # names=[\"drug\", \"accession\", \"drug_name\", \"cas\", \"unii\", \"synonym\", \"key\"],\n",
    ")\n",
    "\n",
    "# Drug.head(0)\n",
    "\n",
    "ChemRepresentation.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### NetworkX Graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[16:33:17 +0530] [INFO]: Graph 'NeuThera' exists.\n",
      "[16:33:17 +0530] [INFO]: Default node type set to 'NeuThera_node'\n"
     ]
    }
   ],
   "source": [
    "G = nxadb.MultiGraph(name=\"NeuThera\")\n",
    "\n",
    "db = G.db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "adding nodes to drug\n",
      "adding nodes to gene\n",
      "adding nodes to disease_doid\n",
      "adding nodes to disease_mesh\n",
      "adding nodes to disease_omim\n"
     ]
    }
   ],
   "source": [
    "# Adding Nodes\n",
    "\n",
    "collections = [\"drug\", \"gene\", \"disease_doid\", \"disease_mesh\", \"disease_omim\"]\n",
    "\n",
    "# Create collections if they don't exist\n",
    "for col in collections:\n",
    "    if not db.has_collection(col):\n",
    "        db.create_collection(col)\n",
    "\n",
    "def add_nodes(df, label, key_column):\n",
    "    print(f\"adding nodes to {label}\")\n",
    "    G.add_nodes_from(\n",
    "        (\n",
    "            f\"{label}/{str(row[key_column]) if pd.notna(row[key_column]) else 'NotAvailable'}\",  # Convert _key to string, replace NaN\n",
    "            row.fillna(\"NaN\").drop(key_column).to_dict()  # Replace NaN in attributes\n",
    "        )\n",
    "        for _, row in df.iterrows()\n",
    "    )\n",
    "    \n",
    "add_nodes(Drug, \"drug\", \"drug\")\n",
    "add_nodes(Gene, \"gene\", \"gene\")\n",
    "add_nodes(Doid, \"disease_doid\", \"doid\")\n",
    "add_nodes(Mesh, \"disease_mesh\", \"mesh\")\n",
    "add_nodes(Omim, \"disease_omim\", \"omim\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "adding Edges to drug-drug\n",
      "adding Edges to drug-gene\n",
      "adding Edges to disease-drug\n",
      "adding Edges to disease-disease\n",
      "adding Edges to disease-function\n",
      "adding Edges to disease-gene\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:urllib3.connectionpool:Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPConnection object at 0x454bcf8d0>: Failed to establish a new connection: [Errno 61] Connection refused')': /_db/NeuThera/_api/document/disease-gene?returnNew=0&silent=0&overwrite=0&returnOld=0\n",
      "WARNING:urllib3.connectionpool:Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPConnection object at 0x454bd0950>: Failed to establish a new connection: [Errno 61] Connection refused')': /_db/NeuThera/_api/document/disease-gene?returnNew=0&silent=0&overwrite=0&returnOld=0\n",
      "WARNING:urllib3.connectionpool:Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPConnection object at 0x454bd17d0>: Failed to establish a new connection: [Errno 61] Connection refused')': /_db/NeuThera/_api/document/disease-gene?returnNew=0&silent=0&overwrite=0&returnOld=0\n",
      "WARNING:urllib3.connectionpool:Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPConnection object at 0x454bd38d0>: Failed to establish a new connection: [Errno 61] Connection refused')': /_db/NeuThera/_api/document/disease-gene?returnNew=0&silent=0&overwrite=0&returnOld=0\n",
      "WARNING:urllib3.connectionpool:Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPConnection object at 0x45c6d4a90>: Failed to establish a new connection: [Errno 61] Connection refused')': /_db/NeuThera/_api/document/disease-gene?returnNew=0&silent=0&overwrite=0&returnOld=0\n",
      "IOStream.flush timed out\n",
      "WARNING:urllib3.connectionpool:Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPConnection object at 0x454bd2690>: Failed to establish a new connection: [Errno 61] Connection refused')': /_db/NeuThera/_api/document/disease-gene?returnNew=0&silent=0&overwrite=0&returnOld=0\n"
     ]
    },
    {
     "ename": "ConnectionAbortedError",
     "evalue": "Can't connect to host(s) within limit (3)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mConnectionAbortedError\u001b[0m                    Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[51], line 33\u001b[0m\n\u001b[1;32m     31\u001b[0m add_edges(DD, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdisease_doid\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdisease_doid\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdoid\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdoid_target\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdisease-disease\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     32\u001b[0m add_edges(DF, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdisease_mesh\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgo\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmesh\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgo\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdisease-function\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m---> 33\u001b[0m \u001b[43madd_edges\u001b[49m\u001b[43m(\u001b[49m\u001b[43mDG\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mdisease_mesh\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mgene\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmesh\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mgene\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mdisease-gene\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m     34\u001b[0m add_edges(FF, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgo\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgo\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgo\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgo_target\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfunction\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     35\u001b[0m add_edges(GF, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgo\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgene\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgo\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgene\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfunction\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "Cell \u001b[0;32mIn[51], line 26\u001b[0m, in \u001b[0;36madd_edges\u001b[0;34m(df, src_label, dst_label, src_col, dst_col, edge_collection)\u001b[0m\n\u001b[1;32m     19\u001b[0m     edges\u001b[38;5;241m.\u001b[39mappend({\n\u001b[1;32m     20\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m_from\u001b[39m\u001b[38;5;124m'\u001b[39m: src_node,\n\u001b[1;32m     21\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m_to\u001b[39m\u001b[38;5;124m'\u001b[39m: dst_node,\n\u001b[1;32m     22\u001b[0m         \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mattributes\n\u001b[1;32m     23\u001b[0m     })\n\u001b[1;32m     25\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m edges:\n\u001b[0;32m---> 26\u001b[0m     \u001b[43mdb\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcollection\u001b[49m\u001b[43m(\u001b[49m\u001b[43medge_collection\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43minsert_many\u001b[49m\u001b[43m(\u001b[49m\u001b[43medges\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages/arango/collection.py:1844\u001b[0m, in \u001b[0;36mCollection.insert_many\u001b[0;34m(self, documents, return_new, sync, silent, overwrite, return_old, overwrite_mode, keep_none, merge, refill_index_caches, version_attribute)\u001b[0m\n\u001b[1;32m   1840\u001b[0m             results\u001b[38;5;241m.\u001b[39mappend(DocumentInsertError(sub_resp, request))\n\u001b[1;32m   1842\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m results\n\u001b[0;32m-> 1844\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_execute\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mresponse_handler\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages/arango/api.py:74\u001b[0m, in \u001b[0;36mApiGroup._execute\u001b[0;34m(self, request, response_handler)\u001b[0m\n\u001b[1;32m     63\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21m_execute\u001b[39m(\n\u001b[1;32m     64\u001b[0m     \u001b[38;5;28mself\u001b[39m, request: Request, response_handler: Callable[[Response], T]\n\u001b[1;32m     65\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Result[T]:\n\u001b[1;32m     66\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Execute an API.\u001b[39;00m\n\u001b[1;32m     67\u001b[0m \n\u001b[1;32m     68\u001b[0m \u001b[38;5;124;03m    :param request: HTTP request.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     72\u001b[0m \u001b[38;5;124;03m    :return: API execution result.\u001b[39;00m\n\u001b[1;32m     73\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> 74\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_executor\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexecute\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mresponse_handler\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages/arango/executor.py:66\u001b[0m, in \u001b[0;36mDefaultApiExecutor.execute\u001b[0;34m(self, request, response_handler)\u001b[0m\n\u001b[1;32m     57\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mexecute\u001b[39m(\u001b[38;5;28mself\u001b[39m, request: Request, response_handler: Callable[[Response], T]) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m T:\n\u001b[1;32m     58\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Execute an API request and return the result.\u001b[39;00m\n\u001b[1;32m     59\u001b[0m \n\u001b[1;32m     60\u001b[0m \u001b[38;5;124;03m    :param request: HTTP request.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     64\u001b[0m \u001b[38;5;124;03m    :return: API execution result.\u001b[39;00m\n\u001b[1;32m     65\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> 66\u001b[0m     resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_conn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend_request\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     67\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m response_handler(resp)\n",
      "File \u001b[0;32m~/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages/arango/connection.py:311\u001b[0m, in \u001b[0;36mBasicConnection.send_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    303\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Send an HTTP request to ArangoDB server.\u001b[39;00m\n\u001b[1;32m    304\u001b[0m \n\u001b[1;32m    305\u001b[0m \u001b[38;5;124;03m:param request: HTTP request.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    308\u001b[0m \u001b[38;5;124;03m:rtype: arango.response.Response\u001b[39;00m\n\u001b[1;32m    309\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    310\u001b[0m host_index \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_host_resolver\u001b[38;5;241m.\u001b[39mget_host_index()\n\u001b[0;32m--> 311\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mprocess_request\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhost_index\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mauth\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_auth\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Documents/Projects/Hackathons/GDG-25/env/lib/python3.11/site-packages/arango/connection.py:178\u001b[0m, in \u001b[0;36mBaseConnection.process_request\u001b[0;34m(self, host_index, request, auth)\u001b[0m\n\u001b[1;32m    175\u001b[0m         host_index \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_host_resolver\u001b[38;5;241m.\u001b[39mget_host_index(indexes_to_filter)\n\u001b[1;32m    176\u001b[0m         tries \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[0;32m--> 178\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mConnectionAbortedError\u001b[39;00m(\n\u001b[1;32m    179\u001b[0m     \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCan\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mt connect to host(s) within limit (\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_host_resolver\u001b[38;5;241m.\u001b[39mmax_tries\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m)\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    180\u001b[0m )\n",
      "\u001b[0;31mConnectionAbortedError\u001b[0m: Can't connect to host(s) within limit (3)"
     ]
    }
   ],
   "source": [
    "edge_collections = [\"drug-drug\", \"drug-gene\", \"disease-drug\", \"disease-disease\", \"disease-function\", \"disease-gene\", \"function-function\", \"gene-function\", \"gene-protien\"]\n",
    "\n",
    "for edge_col in edge_collections:\n",
    "    if not db.has_collection(edge_col):\n",
    "        db.create_collection(edge_col, edge=True)\n",
    "\n",
    "# def add_edges(df, src_label, dst_label, src_col, dst_col, edge_collection):\n",
    "#     print(f\"adding Edges to {edge}\")\n",
    "#     edges = []\n",
    "#     for _, row in df.iterrows():\n",
    "#         src_node = f\"{src_label}/{str(row[src_col]) if pd.notna(row[src_col]) else 'NotAvailable'}\"\n",
    "#         dst_node = f\"{dst_label}/{str(row[dst_col]) if pd.notna(row[dst_col]) else 'NotAvailable'}\"\n",
    "\n",
    "#         attributes = row.fillna(\"NaN\").drop([src_col, dst_col]).to_dict()\n",
    "\n",
    "#         if not isinstance(attributes, dict):\n",
    "#             raise ValueError(\"Attributes must be a dictionary.\")\n",
    "        \n",
    "        # edges.append({\n",
    "        #     '_from': src_node,\n",
    "        #     '_to': dst_node,\n",
    "        #     **attributes\n",
    "        # })\n",
    "    \n",
    "#     G.add_edges_from(edges, collection=edge_collection)\n",
    "\n",
    "def add_edges(df, src_label, dst_label, src_col, dst_col, edge_collection):\n",
    "    \"\"\"Adds edges between nodes, ensuring the correct edge collection is used.\"\"\"\n",
    "    G.add_edges_from(\n",
    "        (\n",
    "            f\"{src_label}/{str(row[src_col]) if pd.notna(row[src_col]) else 'NotAvailable'}\",\n",
    "            f\"{dst_label}/{str(row[dst_col]) if pd.notna(row[dst_col]) else 'NotAvailable'}\",\n",
    "            row.fillna(\"NaN\").drop([src_col, dst_col]).to_dict(),  # Replace NaN in attributes\n",
    "\n",
    "        )\n",
    "        for _, row in df.iterrows()\n",
    "    )\n",
    "\n",
    "add_edges(ChCh, \"drug\", \"drug\", \"drug\", \"drug_target\", \"drug-drug\")\n",
    "add_edges(ChG, \"drug\", \"gene\", \"drug\", \"gene\", \"drug-gene\")\n",
    "add_edges(DCh, \"disease_mesh\", \"drug\", \"mesh\", \"drug\", \"disease-drug\")\n",
    "add_edges(DD, \"disease_doid\", \"disease_doid\", \"doid\", \"doid_target\", \"disease-disease\")\n",
    "add_edges(DF, \"disease_mesh\", \"go\", \"mesh\", \"go\", \"disease-function\")\n",
    "add_edges(DG, \"disease_mesh\", \"gene\", \"mesh\", \"gene\", \"disease-gene\")\n",
    "add_edges(FF, \"go\", \"go\", \"go\", \"go_target\", \"function-function\")\n",
    "add_edges(GF, \"go\", \"gene\", \"go\", \"gene\", \"gene-function\")\n",
    "add_edges(GP, \"gene\", \"protein\", \"ensg\", \"ensp\", \"gene-protien\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Networkx -> ArangoDB -> Gen AI -> x10 Drug Candidates (Novel Drugs SMILE) -> DeepPurpose -> Score -> Best gets selected"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
